---
title: "Introduction to biomartr"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, echo = FALSE, message = FALSE}
options(width = 750)
knitr::opts_chunk$set(
  comment = "#>",
  error = FALSE,
  tidy = FALSE)
```

## Getting Started

A major problem in bioinformatics research is consistent data retrieval. 
The `biomartr` package therefore, aims to provide users with easy to use and diverse interfaces to well curated biological databases such as:

* [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/)
* [BioMart](http://www.biomart.org/)
* [DAVID](http://david.abcc.ncifcrf.gov/)
* [Gene Ontology](http://geneontology.org/)
* [Kegg](http://www.genome.jp/kegg/kegg1a.html)


The collection of functions implemented in `biomartr` enable fast functional annotation queries for a set of genes or entire genomes, biological pathway analyses for a set of genes, biological sequence retrieval, and test statistics for quantifying the significance of enriched gene sets.

Users of the [myTAI](http://cran.r-project.org/web/packages/myTAI/index.html) and [orthologr](https://github.com/HajkD/orthologr) packages will find additional functions that will allow them to integrate `biomartr` to any workflow implemented using `myTAI` or `orthologr` and therefore, allows the highest degree of reproducible phylotranscirptomics research from data retrieval to data visualization.


## Workflow Design

The `bopmartr` package is designed to cover four branches of biological data retrieval:

- __Biological Sequence Retrieval__
- __Functional Annotation Retrieval__
- __Enrichment Analyses__
- __Taxonomic Information Retrieval__

Hence, all functions implemented in `biomartr` serve to fulfill the above listed tasks.



## Biological Sequence Retrieval

In the post-genomic era, biological sequences are used to investigate most phenomena of molecular biology. The growing number of databases and their entries allows us to design meta-studies in a new dimension and to re-investigate known phenomena from a new perspective. Neverthless, from a data science point of view this vast amount of heterogenous data, coming from very different data resources and data standards is very hard to transform from heterogenous to homogenous data. The detection of significant patterns within meta-analyses, therefore relies on high quality data analysis and data science.

Another aspect is reproducibility. Even in cases where a high degree of data homogeneity is achieved, the aspect of scientific reproducibility adds up to a new layer of complexity. Much effort is now being invested to enable high standards of reprpducibility in data driven sciences (e.g. [ROpenSci](http://ropensci.org)). The `biomartr` package aims to be a part of this data science movement. It's functions implement interfaces (Application Programming Interfaces, short _APIs_) to major databases such as [NCBI](http://www.ncbi.nlm.nih.gov) allowing users to access curated data from a single data source. 

The [Sequence Retrieval](https://github.com/HajkD/biomartr/blob/master/vignettes/Sequence_Retrieval.Rmd) vignette will introduce you to the process of biological sequence retrieval using `biomartr`. All functions were designed to allow users the highest (yet possible) degree of reproducibility and transparency for their own analyses.



## Functional Annotation Retrieval

Most phenomena in molecular biology are based on the interaction of genes and their products (proteins). So in case patterns can be found by investigating biological sequences, in most cases it is of interest to map this pattern to a biological function. The [Gene Ontology](http://geneontology.org/) and [Kegg](http://www.genome.jp/kegg/kegg1a.html) consortia provide extensive information on gene functions and pathway memberships.

The `biomartr` packages provides powerful interfaces to functional annotation for specific genes.
Here, the [Functional Annotation](https://github.com/HajkD/biomartr/blob/master/vignettes/Functional_Annotation.Rmd) vignette introduces you to the process of functional information retrieval. 


## Enrichment Analyses

Successfully mapping pattern to functions is the first analytical step in biological research. The next step is to statistically quantify the significance of the observed pattern. 
Enrichment Analyses statistically quantify the significance of gene functions that correspond to an observed pattern.

The `biomartr` package provides useful interfaces to gene enrichment analyses to enable the statistical validation of oberved pattern. 


## Taxonomic Information Retrieval

The evolutionary relationship between species is an integral part of molecular biology research. Genes and metabolic pathways share common descent and therefore, correlating pattern observed in one species with other species from the same taxonomic order enable us to study the origin of observed pattern in a comparative manner. 

The [Taxonomy](https://github.com/HajkD/biomartr/blob/master/vignettes/Taxonomy.Rmd) vignette will introduce you to the retrieval of taxonomic information for molecular biology research.








