---
title: "Sequence Retrieval"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Sequence Retrieval}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, echo = FALSE, message = FALSE}
options(width = 750)
knitr::opts_chunk$set(
  comment = "#>",
  error = FALSE,
  tidy = FALSE)
```


## Biological Sequence Retrieval 

The `biomartr` package allows users to retrieve biological sequences in a very simple and intuitive way.

Using `biomartr`, users can retrieve either genomes, proteomes, or CDS data using the specialized functions:

* `getGenome()`
* `getProteome()`
* `getCDS()`
* `getGFF()`


## Getting Started with Sequence Retrieval

First users can check whether or not the genome, proteome, or CDS of their interest is available for download.

Using the scientific name of the organism of interest, users can check whether the 
corresponding genome is available via the `is.genome.available()` function.

Please note that the first time you run this command it might take a while, because
during the initial execution of this function all neccessary information is retrieved from NCBI
and then stored locally. All further runs are then much faster.

```{r,eval=FALSE}
# checking whether or not the Homo sapiens 
# genome is avaialable for download
is.genome.available("Homo sapiens", db = "refseq")
```

```
[1] TRUE
```

By specifying the `details = TRUE` argument, the genome file size as well as 
additional information can be printed to the console.

```{r,eval=FALSE}
# printing details to the console
is.genome.available("Homo sapiens", details = TRUE, db = "refseq")
```

```
     organism_name  kingdoms   group subgroup file_size_MB chrs organelles plasmids bio_projects
7366  Homo sapiens Eukaryota Animals  Mammals      3236.34   48          1       NA           44
```

Users will observe that the `Homo sapiens` genome file has a size of `3236.34 MB`. 
The argument `db` specifies from which database (`refseq` or `genbank`) organism
information shall be retrieved.

Users can determine the total number of available genomes using the `listGenomes()` function.

```{r,eval=FALSE}
length(listGenomes())
```

```
[1] 17461
```

Hence, currently 17461 genomes (including all kingdoms of life) are stored on [NCBI](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt) servers.


Optionally, users can also specify the database for which the availability of organisms shall be checked.

```{r,eval=FALSE}
# cheking whether Homo sapiens is available in the refseq database
is.genome.available("Homo sapiens", db = "refseq")
```

```
[1] TRUE
```

Or for `genbank`:

```{r,eval=FALSE}
# cheking whether Homo sapiens is available in the genbank database
is.genome.available("Homo sapiens", db = "genbank")
```

```
[1] TRUE
```

```{r,eval=FALSE}
# printing details to the console
is.genome.available("Homo sapiens", details = TRUE, db = "genbank")
```


### Using `is.genome.available()` with ENSEMBL

Users can also specify `db = "ensembl"` to retrieve available organisms provided by ENSEMBL.

```{r,eval=FALSE}
# cheking whether Homo sapiens is available in the ENSEMBL database
is.genome.available("Homo sapiens", db = "ensembl")
```

```
[1] TRUE
```

```{r,eval=FALSE}
# cheking whether Homo sapiens is available in the ENSEMBL database
is.genome.available("Homo sapiens", db = "ensembl", details = TRUE)
```

```
  division taxon_id         name release display_name        accession common_name
1  Ensembl     9606 homo_sapiens      86        Human GCA_000001405.22       human
                                                                    aliases
1 homo, homo sapiens, h_sapiens, enshs, human, hsap, 9606, homsap, hsapiens
                                                       groups assembly
1 core, cdna, vega, otherfeatures, rnaseq, variation, funcgen   GRCh38
```

Please note that the detailed information provided by ENSEMBL differs from the information provided by NCBI. 


The simplest way to work with `listGenomes()` is to print available genomes to the console.

```{r,eval=FALSE}
# the simplest way to retrieve names of available genomes stored within NCBI databases
head(listGenomes() , 5)
```

```
[1] "'Chrysanthemum coronarium' phytoplasma"         
[2] "'Deinococcus soli' Cha et al. 2014"             
[3] "'Echinacea purpurea' witches'-broom phytoplasma"
[4] "Abaca bunchy top virus"                         
[5] "Abalone herpesvirus Victoria/AUS/2009"
```

In case users are interested in a detailed output of the corresponding organism file
stored on NCBI, again they can specify the `details = TRUE` argument.

```{r,eval=FALSE}
# show all details
head(listGenomes(details = TRUE) , 5)
```

```
                                    organism_name kingdoms                       group
1          'Chrysanthemum coronarium' phytoplasma Bacteria         Terrabacteria group
2              'Deinococcus soli' Cha et al. 2014 Bacteria         Terrabacteria group
3 'Echinacea purpurea' witches'-broom phytoplasma Bacteria         Terrabacteria group
4                          Abaca bunchy top virus  Viruses               ssDNA viruses
5           Abalone herpesvirus Victoria/AUS/2009  Viruses dsDNA viruses, no RNA stage
             subgroup file_size_MB chrs organelles plasmids bio_projects
1         Tenericutes     0.739592   NA         NA       NA            1
2 Deinococcus-Thermus     3.236980    1         NA       NA            1
3         Tenericutes     0.545427   NA         NA       NA            1
4         Nanoviridae     0.006422    6         NA       NA            1
5        unclassified     0.211518    1         NA       NA            1
```

Users will observe that the detailed information output includes the `organism_name`, `kingdom`, `group`, `subgroup`, `file_size_MB`,
`chrs`, `organelles`, `plasmids`, and `bio_projects`.

In case users are interested in organisms classified into a specific kingdom of life, they can use the `kingdom`
argument to filter for organisms that are classified into the corresponding kingdom.

```{r,eval=FALSE}
# show all details only for Bacteria
head(listGenomes(kingdom = "Bacteria", details = TRUE) , 5)
```

```
                                    organism_name kingdoms               group
1          'Chrysanthemum coronarium' phytoplasma Bacteria Terrabacteria group
2              'Deinococcus soli' Cha et al. 2014 Bacteria Terrabacteria group
3 'Echinacea purpurea' witches'-broom phytoplasma Bacteria Terrabacteria group
4                           Abiotrophia defectiva Bacteria Terrabacteria group
5                          Acaricomes phytoseiuli Bacteria Terrabacteria group
             subgroup file_size_MB chrs organelles plasmids bio_projects
1         Tenericutes     0.739592   NA         NA       NA            1
2 Deinococcus-Thermus     3.236980    1         NA       NA            1
3         Tenericutes     0.545427   NA         NA       NA            1
4          Firmicutes     2.043440   NA         NA       NA            1
5      Actinobacteria     2.419520   NA         NA       NA            1
```

The following filters can be specified for the `kingdom` argument: `all`, `Archaea`, `Bacteria`, `Eukaryota`, `Viroids`, and `Viruses`.

Furthermore, users can simply count the kingdom specific availability of genomes as following:

```{r,eval=FALSE}
# the number of genomes available for each kingdom
ncbi_genomes <- listGenomes(details = TRUE)
table(ncbi_genomes[ , "kingdoms"])
```

```
 Archaea  Bacteria Eukaryota   Viroids   Viruses 
      586      8932      2029        46      5868
```

Analogous computations can be performed for `group`, `subgroup`, etc.

```{r,eval=FALSE}
# the number of genomes available for each group
ncbi_genomes <- listGenomes(details = TRUE)
table(ncbi_genomes[ , "group"])
```

```
                  Acidobacteria                         Animals 
                             30                             617 
                      Aquificae                   Avsunviroidae 
                             17                               4 
                    Caldiserica                  Chrysiogenetes 
                              2                               2 
                Deferribacteres                      Deltavirus 
                              6                               1 
                    Dictyoglomi                     DPANN group 
                              2                              30 
    dsDNA viruses, no RNA stage                   dsRNA viruses 
                           2405                             244 
                  Elusimicrobia           environmental samples 
                              3                               7 
                  Euryarchaeota                       FCB group 
                            322                             690 
                          Fungi                    Fusobacteria 
                            662                              31 
Nitrospinae/Tectomicrobia group                     Nitrospirae 
                              3                              19 
                          Other                          Plants 
                             18                             179 
                  Pospiviroidae                  Proteobacteria 
                             36                            2718 
                       Protists                       PVC group 
                            190                             113 
     Retro-transcribing viruses                      Satellites 
                            134                             223 
                   Spirochaetes                   ssDNA viruses 
                             84                             848 
                  ssRNA viruses                   Synergistetes 
                           1389                              23 
                     TACK group             Terrabacteria group 
                            130                            3097 
          Thermodesulfobacteria                     Thermotogae 
                              9                              28 
             unassigned viruses            unclassified Archaea 
                             11                              29 
  unclassified archaeal viruses           unclassified Bacteria 
                              3                            1039 
            unclassified phages            unclassified viroids 
                             29                               8 
        unclassified virophages            unclassified viruses 
                              6                              71 
```

Users can also order organisms by their file size.

```{r,eval=FALSE}
# order by file size
library(dplyr)

ncbi_genomes <- listGenomes(details = TRUE)
head(arrange(ncbi_genomes, desc(file_size_MB)) , 10)
```

```
            organism_name  kingdoms   group    subgroup file_size_MB chrs organelles
1       Pinus lambertiana Eukaryota  Plants Land Plants     27602.70   NA         NA
2            Picea glauca Eukaryota  Plants Land Plants     24627.00   NA          1
3             Pinus taeda Eukaryota  Plants Land Plants     22061.90   NA         NA
4   Pseudotsuga menziesii Eukaryota  Plants Land Plants     14673.20   NA         NA
5      Locusta migratoria Eukaryota Animals     Insects      5759.80   NA         NA
6        Orycteropus afer Eukaryota Animals     Mammals      4444.08   NA          1
7  Chrysochloris asiatica Eukaryota Animals     Mammals      4210.11   NA          1
8   Elephantulus edwardii Eukaryota Animals     Mammals      3843.98   NA         NA
9     Apodemus sylvaticus Eukaryota Animals     Mammals      3758.14   NA         NA
10        Triticum urartu Eukaryota  Plants Land Plants      3747.05   NA         NA
   plasmids bio_projects
1        NA            1
2        NA            2
3        NA            1
4        NA            1
5        NA            1
6        NA            1
7        NA            1
8        NA            1
9        NA            1
10       NA            1
```

This analysis shows that `Pinus lambertiana` has the largest genome available on the NCBI server.


Internally, the `listGenomes()` function downloads the [Genome Reports](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt) file
from NCBI and stores it in a `tempfile()` folder named `_ncbi_downloads/overview.txt`. It is only downloaded once
and is then accessed from your hard drive. In case users would like to update the [Genome Reports](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt) file,
they can specify the `update = TRUE` argument which allows them to reload the [Genome Reports](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt) file from 
the NCBI server.

```{r,eval=FALSE}
# users can also update the organism table using the 'update' argument
head(listGenomes(details = TRUE, update = TRUE) , 5)
```

```
                                    organism_name kingdoms                       group
1          'Chrysanthemum coronarium' phytoplasma Bacteria         Terrabacteria group
2              'Deinococcus soli' Cha et al. 2014 Bacteria         Terrabacteria group
3 'Echinacea purpurea' witches'-broom phytoplasma Bacteria         Terrabacteria group
4                          Abaca bunchy top virus  Viruses               ssDNA viruses
5           Abalone herpesvirus Victoria/AUS/2009  Viruses dsDNA viruses, no RNA stage
             subgroup file_size_MB chrs organelles plasmids bio_projects
1         Tenericutes     0.739592   NA         NA       NA            1
2 Deinococcus-Thermus     3.236980    1         NA       NA            1
3         Tenericutes     0.545427   NA         NA       NA            1
4         Nanoviridae     0.006422    6         NA       NA            1
5        unclassified     0.211518    1         NA       NA            1

```

Again, the `listGenomes()` function can be users to filter for available genome information in refseq.

```{r,eval=FALSE}
# list all Eukaryota that are stored in refseq
head(listGenomes(kingdom = "Eukaryota", db = "refseq") , 20)
```

```
                    organism_name
1               Agaricus bisporus
2           Auricularia subglabra
3            Komagataella phaffii
4           Arthroderma benhamiae
5                Arthroderma otae
6            Aspergillus clavatus
7              Aspergillus flavus
8           Aspergillus fumigatus
9            Aspergillus nidulans
10              Aspergillus niger
11             Aspergillus oryzae
12            Aspergillus terreus
13        Phaeoacremonium minimum
14 Batrachochytrium dendrobatidis
15            Ordospora colligata
16               Bipolaris oryzae
17          Bipolaris sorokiniana
18              Bipolaris zeicola
19               Botrytis cinerea
20               Candida albicans
```

Or analogous:


```{r,eval=FALSE}
# the number of genomes available for each kingdom stored in refseq
ncbi_genomes <- listGenomes(details = TRUE, db = "refseq")
table(ncbi_genomes[ , "kingdoms"])
```

```
 Archaea  Bacteria Eukaryota 
      271      5371       675 
```

__Note__ that when running the `listGenomes()` function for the first time, it might take a while until
the function returns any results, because necessary information need to be downloaded from NCBI databases. All subsequent executions of `listGenomes()`
will then respond very fast, because they will access the corresponding files stored on your hard drive.

## Downloading Biological Sequences and Annotations

After checking for the availability of sequence information for an organism of interest,
the next step is to download the corresponding genome, proteome, or CDS file in `fasta` format.
The following functions allow users to download proteomes, genomes, and CDS files from several
database resources such as: `refseq`. When a corresponding proteome, genome, or CDS file was
loaded to your hard-drive, a documentation `*.txt` file is generated storing `File Name`, `Organism`,
`Database`, `URL`, `DATE`, `assembly_accession`, `bioproject`, `biosample`, `taxid`, `version_status`, `release_type`,
`seq_rel_date` etc. information. This way a better reproducibility of proteome, genome, and CDS versions
used for subsequent data analyses can be achieved.


### Genome Retrieval

The easiest way to download a genome is to use the `getGenome()` function.

In this example we will download the genome of `Homo sapiens`.

The `getGenome()` function is an interface function to the [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/) or [NCBI genbank](http://www.ncbi.nlm.nih.gov/genbank/about/) databases from
which corresponding genomes can be retrieved.

For this purpose users need to specify the kingdom in which their organism of interest is classified into,
e.g. `"archaea"`,`"bacteria"`, `"fungi"`, `"invertebrate"`, `"plant"`, `"protozoa"`, `"vertebrate_mammalian"`, or `"vertebrate_other"` (see also `?getKingdoms`)
and then the scientific name of the organism of interest.

```{r,eval=FALSE}
# download the genome of Homo sapiens from refseq
# and store the corresponding genome file in '_ncbi_downloads/genomes'
getGenome( db       = "refseq", 
           organism = "Homo sapiens",
           path     = file.path("_ncbi_downloads","genomes") )
```

The `getGenome()` function creates a directory named `'_ncbi_downloads/genomes'` into which the corresponding genome named `GCF_000001405.34_GRCh38.p8_genomic.fna.gz` is downloaded. The return value of `getGenome()` is the path to the genome
that can be used as input to the `read_genome()` function. The `read_genome()` function reads the genome and stores it as as `data.table` object.

```{r,eval=FALSE}
# read genome as data.table object
Human_Genome <- read_genome(getGenome( db       = "refseq", 
                                       organism = "Homo sapiens",
                                       path     = file.path("_ncbi_downloads","genomes") ), 
                            format = "fasta")

```

In case users would like to store the genome file at a different location,
they can specify the `path = file.path("put","your","path","here")` argument.

In case, users wish to download genomes from [NCBI genbank](http://www.ncbi.nlm.nih.gov/genbank/about/) instead of [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/), they can specify the
argument `db = "genbank"` in `getGenome()`.

### Proteome Retrieval

The `getProteome()` function is also an interface function to the [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/) or [NCBI genbank](http://www.ncbi.nlm.nih.gov/genbank/about/) databases from
which corresponding genomes can be retrieved. It works analogous to `getGenome()`. 


```{r,eval=FALSE}
# download the proteome of Homo sapiens from refseq
# and store the corresponding proteome file in '_ncbi_downloads/proteomes'
getProteome( db       = "refseq", 
             organism = "Homo sapiens",
             path     = file.path("_ncbi_downloads","proteomes") )
```

The `getProteome()` function creates a directory named `_ncbi_downloads/proteomes` into which the orresponding proteome named `GCF_000001405.34_GRCh38.p8_protein.faa.gz` is downloaded. TThe return value of `getProteome()` is the path to the proteome
that can be used as input to the `read_proteome()` function. The `read_proteome()` function reads the proteome and stores it as as `data.table` object.


```{r,eval=FALSE}
# read proteome as data.table object
Human_Proteome <- read_proteome(getProteome( db       = "refseq", 
                                             organism = "Homo sapiens",
                                             path     = file.path("_ncbi_downloads","proteomes") ), 
                                format = "fasta")
```

In case users would like to store the proteome file at a different location,
they can specify the `path = file.path("put","your","path","here")` argument.

Or using the pipeline logic of the [magrittr](https://github.com/smbache/magrittr) package:

```{r,eval=FALSE}
library(magrittr)
# read proteome as data.table object
Human_Proteome <- getProteome( db       = "refseq", 
                               organism = "Homo sapiens",
                               path     = file.path("_ncbi_downloads","proteomes")) %>%
    read_proteome(format = "fasta")
```

In case, users wish to download genomes from [NCBI genbank](http://www.ncbi.nlm.nih.gov/genbank/about/) instead of [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/), they can specify the
argument `db = "genbank"` in `getProteome()`. In case users select `db = "genbank"`, the `format` argument in `read_proteome()` must be adapted accordingly to
handle genbank formatted files as input: `read_proteome(..., format = "gbk")`.

```{r,eval=FALSE}
# read proteome as data.table object
Human_Proteome <- read_proteome(getProteome( db       = "genbank", 
                                             organism = "Homo sapiens",
                                             path     = file.path("_ncbi_downloads","proteomes") ), 
                                format = "gbk")
```


### CDS Retrieval

The `getCDS()` function is also an interface function to the [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/) database from
which the corresponding CDS files are downloaded. It works analogous to the `getGenome()` and `getProteome()` functions but for CDS files. 


```{r,eval=FALSE}
# download the genome of Homo sapiens from refseq
# and store the corresponding genome CDS file in '_ncbi_downloads/CDS'
getCDS( db       = "refseq", 
        organism = "Homo sapiens",
        path     = file.path("_ncbi_downloads","CDS") )
```

The `getCDS()` function creates a directory named `_ncbi_downloads/CDS` into which
corresponding CDS are loaded. The `read_cds()` function allows users to
read the correspondning CDS as `data.table` object.

```{r,eval=FALSE}
# read CDS as data.table object
Human_cds <- read_cds(getCDS( db       = "refseq", 
                              organism = "Homo sapiens",
                              path     = file.path("_ncbi_downloads","CDS") ), 
                      format = "fasta")
```

In case users would like to store the CDS file at a different location,
they can specify the `path = file.path("put","your","path","here")` argument.


Furthermore, the `getCDS()` function checks whether all CDS sequences can be divided by 3 (codons).
In case sequences of particualar genes cannot be divided by 3, a warning massage is returned to quantify the
number of corresponding genes. In case users would like to extract these sequences from their data, they can specify the
`delete_delete_corrupt = TRUE` argument, which will then delete all corrupt CDS sequences.


### Retrieve the annotation file of a particular genome

Finally, users can download the annotation `*.gff` file for a particular genome of interest using the `getGFF()` function.
In `getGFF()`, users again can specify the database from which annotation files shall be retrieved.

Example `RefSeq`:

```{r,eval=FALSE}
# download the annotation file of Homo sapiens from refseq
# and store the corresponding file in '_ncbi_downloads/annotation'
getGFF( db       = "refseq", 
        organism = "Homo sapiens", 
        path = file.path("_ncbi_downloads","annotation"))
```

Example `GenBank`:

```{r,eval=FALSE}
# download the annotation file of Homo sapiens from genbank
# and store the corresponding file in '_ncbi_downloads/annotation'
getGFF( db       = "genbank", 
        organism = "Homo sapiens", 
        path = file.path("_ncbi_downloads","annotation"))
```


Example `Ensembl`:

```{r,eval=FALSE}
# download the annotation file of Homo sapiens from refseq
# and store the corresponding file in '_ncbi_downloads/annotation'
getGFF( db       = "ensembl", 
        organism = "Homo sapiens", 
        path = file.path("_ncbi_downloads","annotation"))
```

The `getGFF()` function retrieves `*.gff` annotation files of an organism of interest and stores the annotation file in a specified folder, e.g. `file.path("_ncbi_downloads","annotation")`.
The return value of `getGFF()` is the file path to the downloaded annotation file. Internally, again a documentation file is generated to log the origin, date, information, etc. of the retrieved 
`*.gff` file.

Taken together, `getGFF()` in combination with `getGenome()`, `getProteome()` and `getCDS()` allows users to retrieve the genome information together with the
corresponding `*.gff` annotation file to make sure that they both have the same version and origin.





