---
title: "Sequence Retrieval"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Sequence Retrieval}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, echo = FALSE, message = FALSE}
options(width = 750)
knitr::opts_chunk$set(
  comment = "#>",
  error = FALSE,
  tidy = FALSE)
```


The `biomartr` package allows you to retrieve biological sequences in very simple and intuitive way.

Using `biomartr`, you can retrieve either genomes, proteomes, or CDS data using the specialized functions:

* `getGenome()`
* `getProteome()`
* `getCDS()`
* `geneSequence()`


## Getting Started with Sequence Retrieval

First you can check whether the genome, proteome, or CDS is available for download.

Using the scientific name of the organism of interest, you can check whether the 
corresponding genome is available via the `is.genome.available()` function.

```{r,eval=FALSE}
# checking whether or not the Arabidopsis thaliana 
# genome is avaialable for download
is.genome.available("Arabidopsis thaliana")


```

```
[1] TRUE
```

By specifying the `details = TRUE` argument, the genome file size as well as 
additional information can be proted to the console.

```{r,eval=FALSE}
# printing details to the console
is.genome.available("Arabidopsis thaliana", details = TRUE)

```

```
           organism_name  kingdoms  group    subgroup file_size_MB chrs organelles plasmids bio_projects
682 Arabidopsis thaliana Eukaryota Plants Land Plants      119.668    6          2       NA            6

```

As you can see, the `Arabidopsis thaliana` genome file has a size of `119.668 MB`.

__Note:__ The availability of genomes has been taken from [NCBI](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt).

You can determine the total number of available genomes using the `listGenomes()` function.

```{r,eval=FALSE}

length(listGenomes())

```

```
[1] 10470
```

Hence, currently 10470 genomes (including all kingdoms of life) are stored on [NCBI](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt) servers.


Optionally, you can also specify the database for which the availability of organisms shall be checked.

```{r,eval=FALSE}
# cheking whether A. thaliana is available in the refseq database
is.genome.available("Arabidopsis thaliana", database = "refseq")

```

```
[1] TRUE
```

You can also determien the total number of genomes stored in refseq.

```{r,eval=FALSE}

length(listGenomes(database = "refseq"))
```

```
[1] 3931
```

This result shows that 3931 genomes are stored in refseq.

The simplest way to work with `listGenomes()` is to print available genomes to the console.

```{r,eval=FALSE}

# the simplest way to retrieve names of available genomes stored within NCBI databases
head(listGenomes() , 5)


```

```
[1] "'Chrysanthemum coronarium' phytoplasma"       "Abaca bunchy top virus"                      
[3] "Abalone herpesvirus Victoria/AUS/2009"        "Abalone shriveling syndrome-associated virus"
[5] "Abelson murine leukemia virus"
```

In case you are interested in a detailed output of the corresponding organism file
stored on NCBI, again you can specify the `details = TRUE` argument.

```{r,eval=FALSE}

# show all details
head(listGenomes(details = TRUE) , 5)

```

```
                               organism_name kingdoms                       group
1       'Chrysanthemum coronarium' phytoplasma Bacteria                 Tenericutes
2                       Abaca bunchy top virus  Viruses               ssDNA viruses
3        Abalone herpesvirus Victoria/AUS/2009  Viruses dsDNA viruses, no RNA stage
4 Abalone shriveling syndrome-associated virus  Viruses dsDNA viruses, no RNA stage
5                Abelson murine leukemia virus  Viruses  Retro-transcribing viruses

      subgroup file_size_MB chrs organelles plasmids bio_projects
1   Mollicutes     0.739592   NA         NA       NA            1
2  Nanoviridae     0.006422    6         NA       NA            1
3 unclassified     0.211518    1         NA       NA            1
4 unclassified     0.034952    1         NA       NA            1
5 Retroviridae     0.005894    1         NA       NA            1
```

As you can see, detailed information includes the `organism_name`, `kingdom`, `group`, `subgroup`, `file_size_MB`,
`chrs`, `organelles`, `plasmids`, and `bio_projects`.

If you are interested in organisms classified into a specific kingdom of life, you can use the `kingdom`
argument to filter for organisms that are classified into the corresponding kingdom.

```{r,eval=FALSE}

# show all details only for Bacteria
head(listGenomes(kingdom = "Bacteria", details = TRUE) , 5)

```

```
                           organism_name kingdoms          group              subgroup file_size_MB
1 'Chrysanthemum coronarium' phytoplasma Bacteria    Tenericutes            Mollicutes     0.739592
2                  Abiotrophia defectiva Bacteria     Firmicutes               Bacilli     2.043440
3                 Acaricomes phytoseiuli Bacteria Actinobacteria        Actinobacteria     2.419520
4                          Acaryochloris Bacteria  Cyanobacteria Oscillatoriophycideae     7.875480
5                   Acaryochloris marina Bacteria  Cyanobacteria Oscillatoriophycideae     8.361600
  chrs organelles plasmids bio_projects
1   NA         NA       NA            1
2   NA         NA       NA            1
3   NA         NA       NA            1
4   NA         NA       NA            1
5    1         NA        9            1

```

The following filters can be selected for `kingdom`: `all`, `Archaea`, `Bacteria`, `Eukaryota`, `Viroids`, and `Viruses`.

Furthermore, you can simply count the kingdom specific availability of genomes as following.

```{r,eval=FALSE}

# the number of genomes available for each kingdom
ncbi_genomes <- listGenomes(details = TRUE)
table(ncbi_genomes[ , "kingdoms"])

```

```
  Archaea  Bacteria Eukaryota   Viroids   Viruses 
      332      4892      1107        44      4095
```

Analogous computations can be performed for `group`, `subgroup`, etc.

```{r,eval=FALSE}

# the number of genomes available for each group
ncbi_genomes <- listGenomes(details = TRUE)
table(ncbi_genomes[ , "group"])

```

```


                   Actinobacteria                   Aenigmarchaeota                           Animals 
                              796                                 3                               403 
                        Aquificae                   Armatimonadetes                     Avsunviroidae 
                               16                                 3                                 4 
     Bacteroidetes/Chlorobi group                       Caldiserica  Chlamydiae/Verrucomicrobia group 
                              422                                 2                                55 
                      Chloroflexi                    Chrysiogenetes                     Crenarchaeota 
                               25                                 2                                56 
                    Cyanobacteria                   Deferribacteres               Deinococcus-Thermus 
                               88                                 6                                40 
                       Deltavirus                    Diapherotrites                       Dictyoglomi 
                                1                                 3                                 2 
      dsDNA viruses, no RNA stage                     dsRNA viruses                     Elusimicrobia 
                             1747                               196                                 1 
                    Euryarchaeota Fibrobacteres/Acidobacteria group                        Firmicutes 
                              212                                44                              1008 
                            Fungi                      Fusobacteria                  Gemmatimonadetes 
                              414                                22                                 5 
                     Korarchaeota                     Nanoarchaeota                 Nanohaloarchaeota 
                                1                                10                                 4 
                      Nitrospinae                       Nitrospirae                             Other 
                                2                                 8                                11 
                    Parvarchaeota                    Planctomycetes                            Plants 
                                2                                19                               138 
                    Pospiviroidae                    Proteobacteria                          Protists 
                               35                              1890                               141 
       Retro-transcribing viruses                        Satellites                      Spirochaetes 
                              123                               205                                77 
                    ssDNA viruses                     ssRNA viruses                     Synergistetes 
                              658                              1092                                18 
                      Tenericutes                    Thaumarchaeota             Thermodesulfobacteria 
                              115                                40                                 7 
                      Thermotogae                unassigned viruses              unclassified Archaea 
                               20                                 8                                 1 
    unclassified archaeal viruses             unclassified Bacteria               unclassified phages 
                                2                               199                                50 
             unclassified viroids           unclassified virophages              unclassified viruses 
                                5                                 5                                 8
```

You can also order organisms by their file size.

```{r,eval=FALSE}

# order by file size
library(dplyr)

ncbi_genomes <- listGenomes(details = TRUE)
head(arrange(ncbi_genomes, desc(file_size_MB)) , 10)

```

```
               organism_name  kingdoms   group      subgroup file_size_MB chrs organelles plasmids
1               Picea glauca Eukaryota  Plants   Land Plants     25471.90   NA         NA       NA
2  Acanthoscurria geniculata Eukaryota Animals Other Animals      7178.40   NA         NA       NA
3         Locusta migratoria Eukaryota Animals       Insects      5759.80   NA          1       NA
4           Orycteropus afer Eukaryota Animals       Mammals      4444.08   NA          1       NA
5     Chrysochloris asiatica Eukaryota Animals       Mammals      4210.11   NA          1       NA
6      Elephantulus edwardii Eukaryota Animals       Mammals      3843.98   NA         NA       NA
7          Triticum aestivum Eukaryota  Plants   Land Plants      3800.33    1          1       NA
8            Triticum urartu Eukaryota  Plants   Land Plants      3747.05   NA          1       NA
9       Dasypus novemcinctus Eukaryota Animals       Mammals      3631.52   NA          1       NA
10         Nicotiana tabacum Eukaryota  Plants   Land Plants      3613.16   NA          1       NA
```

This analysis shows that `Picea glauca` has the largest genome available on the NCBI server.


Internally, the `listGenomes()` function downloads the [Genome Reports](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt) file
from NCBI and stores it in a folder that is being created named `_ncbi_downloads/overview.txt`. It is only downloaded once
and is then being accessed from your hard drive. In case you want to update the [Genome Reports](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt) file,
you can specify the `update = TRUE` argument which allows you to reload the [Genome Reports](ftp://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/overview.txt) file from 
the NCBI server.

```{r,eval=FALSE}

# you can also update the organism table using the 'update' argument
head(listGenomes(details = TRUE, update = TRUE) , 5)

```

```
                                 organism_name kingdoms                       group
1       'Chrysanthemum coronarium' phytoplasma Bacteria                 Tenericutes
2                       Abaca bunchy top virus  Viruses               ssDNA viruses
3        Abalone herpesvirus Victoria/AUS/2009  Viruses dsDNA viruses, no RNA stage
4 Abalone shriveling syndrome-associated virus  Viruses dsDNA viruses, no RNA stage
5                Abelson murine leukemia virus  Viruses  Retro-transcribing viruses

      subgroup file_size_MB chrs organelles plasmids bio_projects
1   Mollicutes     0.739592   NA         NA       NA            1
2  Nanoviridae     0.006422    6         NA       NA            1
3 unclassified     0.211518    1         NA       NA            1
4 unclassified     0.034952    1         NA       NA            1
5 Retroviridae     0.005894    1         NA       NA            1

```

Again, the `listGenomes()` function can be limited to search for available genome information in refseq.

```{r,eval=FALSE}

# list all Eukaryota that are stored in refseq
head(listGenomes(kingdom = "Eukaryota", database = "refseq") , 20)
```

```
                    organism_name
1               Agaricus bisporus
2          Ajellomyces capsulatus
3        Ajellomyces dermatitidis
4           Arthroderma benhamiae
5                Arthroderma otae
6            Aspergillus clavatus
7              Aspergillus flavus
8           Aspergillus fumigatus
9            Aspergillus nidulans
10              Aspergillus niger
11             Aspergillus oryzae
12            Aspergillus terreus
13           Auricularia delicata
14 Batrachochytrium dendrobatidis
15        Baudoinia compniacensis
16               Bipolaris oryzae
17          Bipolaris sorokiniana
18              Bipolaris zeicola
19               Botrytis cinerea
20               Candida albicans
```

Or analogous:


```{r,eval=FALSE}

# the number of genomes available for each kingdom stored in refseq
ncbi_genomes <- listGenomes(details = TRUE, database = "refseq")
table(ncbi_genomes[ , "kingdoms"])

```

```
  Archaea  Bacteria Eukaryota 
      193      3299       439
```

__Note__: when running the `listGenomes()` function for the first time, it might take a while until
the function returns your results, because necessary information are downloaded from databases
and then are being stored in the `_ncbi_downloads` folder on your hard drive. All following executions of `listGenomes()`
will then respond very fast, because they will access the corresponding files stored on your hard drive.

## Downloading Biological Sequences

After checking for the availability of sequence information for an organism of interest,
the next step is to download the corresponding genome, proteome, or CDS file in `fasta` format.


### Genome Retrieval

The easiest way to download a genome is to use the `getGenome()` function.

In this example we will download the genome of `A. thaliana`.

The `getGenome()` function is an interface function to the [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/) database from
which corresponding genomes are downloaded.

For this purpose you need to specify the kingdom in which your organism of interest is classified into,
e.g. `"archaea"`,`"bacteria"`, `"fungi"`, `"invertebrate"`, `"plant"`, `"protozoa"`, `"vertebrate_mammalian"`, or `"vertebrate_other"`
and then the scientific name of the organism of interest.

```{r,eval=FALSE}

# download the genome of Arabidopsis thaliana from refseq
# and store the corresponding genome file in '_ncbi_downloads/genomes'
Ath_genome <- getGenome(db = "refseq", kingdom = "plant",
                        organism = "Arabidopsis thaliana",
                        clean_folder = FALSE)


```

Internally, the `getGenome()` function creates a directory named `_ncbi_downloads/genomes` in which
corresponding genomes are loaded and then sourced as data.table object into the current R session.
When specifying `clean_folder = FALSE`, the `_ncbi_downloads/genomes` folder will not be removed and
the corresponding genome does not need to be downloaded again when sourcing it into the current R session via 
`getGenome()`.


### Proteome Retrieval

The `getProteome()` function is also an interface function to the [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/) database from
which corresponding proteomes are downloaded. It works exactly the same as `getGenome()`. 


```{r,eval=FALSE}

# download the proteome of Arabidopsis thaliana from refseq
# and store the corresponding proteome file in '_ncbi_downloads/proteomes'
Ath_proteome <- getProteome(db = "refseq", kingdom = "plant",
                        organism = "Arabidopsis thaliana",
                        clean_folder = FALSE)


```

Internally, the `getProteome()` function creates a directory named `_ncbi_downloads/proteome` in which
corresponding proteomes are loaded and then sourced as data.table object into the current R session.
When specifying `clean_folder = FALSE`, the `_ncbi_downloads/proteomes` folder will not be removed and
the corresponding proteome does not need to be downloaded again when sourcing it into the current R session via 
`getProteome()`.





### CDS Retrieval

The `getCDS()` function is also an interface function to the [NCBI refseq](http://www.ncbi.nlm.nih.gov/refseq/about/) database from
which corresponding CDS files are downloaded. It works exactly the same as `getGenome()` and `getProteome()` but for CDS files. 


```{r,eval=FALSE}

# download the genome of Arabidopsis thaliana from refseq
# and store the corresponding genome CDS file in '_ncbi_downloads/CDS'
Ath_cds <- getCDS(db = "refseq", kingdom = "plant",
                        organism = "Arabidopsis thaliana",
                        clean_folder = FALSE)


```

Internally, the `getCDS()` function creates a directory named `_ncbi_downloads/CDS` in which
corresponding CDS files are loaded and then sourced as data.table object into the current R session.
When specifying `clean_folder = FALSE`, the `_ncbi_downloads/CDS` folder will not be removed and
the corresponding proteome does not need to be downloaded again when sourcing it into the current R session via 
`getCDS()`.

Furthermore, the `getCDS()` function checks whether all CDS sequences can be divided by 3 (codons).
In case sequences of particualar genes cannot be divided by 3, a warning massage is returned to quantify the
number of correspondning genes. In case you would like to extract these sequences from your data, you can specify the
`delete_delete_corrupt = TRUE` argument, which will then delete all corrupt CDS sequences.


### Retrieving sequences for a set of genes

For most analyses only subsets of sequences (taken from the entire genome) are needed.
This section introduces several approaches to select a set of sequences for furthr analyses.

#### Using the output from `getProteome()`

As you saw before, the `getProteome()` function allows you to download 
the entire proteome of a specific organism of interest that is stored in refseq.

```{r,eval=FALSE}

# download the proteome of Arabidopsis thaliana from refseq
# and store the corresponding proteome file in '_ncbi_downloads/proteomes'
Ath_proteome <- getProteome(db = "refseq", kingdom = "plant",
                        organism = "Arabidopsis thaliana",
                        clean_folder = FALSE)


```

Again, we download the _A. thaliana_ proteome and now are interested in the following two genes for 
subsequent analyses, `AT1G06090` and `AT1G06100`. Both genes are memebers of fatty acid desaturase family and take over functions in oxidoreductase activity ([see details](http://arabidopsis.org/servlets/TairObject?id=136833&type=locus)). To access the protein sequences of these genes in a proteome downloaded from refseq, first you need to map the `tair_locus` id to the corresponding `refseq` ids.

For this purpose the you can use the `biomart()` function [see [Functional Annotation](https://github.com/HajkD/biomartr/blob/master/vignettes/Functional_Annotation.Rmd) for details].


```{r,eval=FALSE}

bm <- biomart(genes = c("AT1G06090", "AT1G06100"),
              mart = "plants_mart_23", dataset = "athaliana_eg_gene",
              attributes = c("refseq_peptide"),
              filters = "tair_locus")
```

```
  filter_input filter_output refseq_peptide        NA
1    AT1G06090     AT1G06100      AT1G06100 NP_172100
2    AT1G06100     AT1G06090      AT1G06090 NP_172099

```

```{r,eval=FALSE}
genes_of_interest <- which(stringr::str_detect(bm[ , 4], Ath_proteome[ , geneids]))

Ath_proteome[ , ]
```









