<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Covid-19: Genomic Data Retrieval and Evolutionary Genomics • biomartr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script>




<meta property="og:title" content="Covid-19: Genomic Data Retrieval and Evolutionary Genomics" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-title-body">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">biomartr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="articles/BioMart_Examples.html">Ensembl BioMart Examples</a>
    </li>
    <li>
      <a href="articles/Database_Retrieval.html">NCBI Database Retrieval</a>
    </li>
    <li>
      <a href="articles/Functional_Annotation.html">Functional Annotation with biomartr</a>
    </li>
    <li>
      <a href="articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a>
    </li>
    <li>
      <a href="articles/Sequence_Retrieval.html">Sequence Retrieval</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ropensci/biomartr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Covid-19: Genomic Data Retrieval and Evolutionary Genomics</h1>
    </div>


<div id="covid-19-genomic-data-retrieval-and-evolutionary-genomics" class="section level1">

<blockquote>
<p>Data is the most powerful weapon humans possess to combat Covid-19.</p>
</blockquote>
<p>In an unprecented scale, the research community came together in the past months to combat Covid-19. The most significant contribution any researcher can do to fight this virus is: data collection and data analytics. <strong>Everyone can and should make a contribution</strong>.</p>
<p>Here, we aim to provide tools to make it easier for virologists and virus bioinformaticians to retrieve and analyse the genomic sequences of all viruses and in particular, Covid-19.</p>
<p>Unfortunately, we are neither virologists and nor virus bioinformaticians, so we are not aware of all available virus resources that would help the community when being able to retrieve and analyse. <strong>Please contact us or open issues on this GitHub page to indicate how we can add further tools or functionality to make the life of virus researchers easier.</strong></p>
<p>Here some useful data analytics resources we found so far:</p>
<ul>
<li>
<a href="https://www.nytimes.com/2020/03/20/health/coronavirus-data-logarithm-chart.html">New York Times article with beautiful data analytics</a>
<ul>
<li>data can be found here: <a href="https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data/csse_covid_19_daily_reports" class="uri">https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data/csse_covid_19_daily_reports</a>
</li>
</ul>
</li>
<li><p><a href="https://towardsdatascience.com/top-5-r-resources-on-covid-19-coronavirus-1d4c8df6d85f">Top 15 R resources on Novel COVID-19 Coronavirus</a></p></li>
<li><p><a href="https://www.reddit.com/r/datasets/comments/exnzrd/coronavirus_datasets/">Coronavirus Datasets</a></p></li>
</ul>
</div>
<div id="retrieval-of-meta-data-and-genomic-sequences" class="section level1">
<h1 class="hasAnchor">
<a href="#retrieval-of-meta-data-and-genomic-sequences" class="anchor"></a>Retrieval of meta-data and genomic sequences</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># retrieve genome, proteome, cds, gff/gtf, rna, repeat masker, assembly stats files for SARS-Cov-2</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co"># alternatively, users can specify the taxid -&gt; 'organism = 2697049' </span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">biomartr<span class="op">::</span><span class="kw"><a href="https://docs.ropensci.org/biomartr/reference/getCollection.html">getCollection</a></span>(<span class="dt">db =</span> <span class="st">"refseq"</span>, </a>
<a class="sourceLine" id="cb1-4" data-line-number="4">                        <span class="dt">organism =</span> <span class="st">"Severe acute respiratory syndrome coronavirus 2"</span>, </a>
<a class="sourceLine" id="cb1-5" data-line-number="5">                        <span class="dt">path =</span> <span class="st">"sars_cov_2"</span>)</a></code></pre></div>
<pre><code>Starting collection retrieval (genome, proteome, cds, gff/gtf, rna, repeat masker, assembly stats) for Severe_acute_respiratory_syndrome_coronavirus_2 ...
Starting genome retrieval of 'Severe acute respiratory syndrome coronavirus 2' from refseq ...


|===========================================================================| 100%   53 MB
Genome download of Severe_acute_respiratory_syndrome_coronavirus_2 is completed!
Checking md5 hash of file: sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt ...
The md5 hash of file 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt' matches!
The genome of 'Severe_acute_respiratory_syndrome_coronavirus_2' has been downloaded to 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2' and has been named 'Severe_acute_respiratory_syndrome_coronavirus_2_genomic_refseq.fna.gz'.


Starting proteome retrieval of 'Severe acute respiratory syndrome coronavirus 2' from refseq ...


Proteome download of Severe_acute_respiratory_syndrome_coronavirus_2 is completed!
Checking md5 hash of file: sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt ...
The md5 hash of file 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt' matches!
The proteome of 'Severe_acute_respiratory_syndrome_coronavirus_2' has been downloaded to 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2' and has been named 'Severe_acute_respiratory_syndrome_coronavirus_2_protein_refseq.faa.gz' .


Starting CDS retrieval of 'Severe acute respiratory syndrome coronavirus 2' from refseq ...


|===========================================================================| 100%   53 MB
CDS download of Severe_acute_respiratory_syndrome_coronavirus_2 is completed!
Checking md5 hash of file: sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt ...
The md5 hash of file 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt' matches!
The genomic CDS of 'Severe_acute_respiratory_syndrome_coronavirus_2' has been downloaded to 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2' and has been named 'Severe_acute_respiratory_syndrome_coronavirus_2_cds_from_genomic_refseq.fna.gz' .


Starting GFF retrieval of 'Severe acute respiratory syndrome coronavirus 2' from refseq ...


|===========================================================================| 100%   53 MB
GFF download of Severe_acute_respiratory_syndrome_coronavirus_2 is completed!
Checking md5 hash of file: sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt ...
The md5 hash of file 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt' matches!
Importing 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_genomic_refseq.gff.gz' ...


Starting RNA retrieval of 'Severe acute respiratory syndrome coronavirus 2' from refseq ...


|===========================================================================| 100%   53 MB
The download session seems to have timed out at the FTP site 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/858/895/GCF_009858895.2_ASM985889v3/GCF_009858895.2_ASM985889v3_rna_from_genomic.fna.gz'. This could be due to an overload of queries to the databases. Please restart this function to continue the data retrieval process or wait for a while before restarting this function in case your IP address was logged due to an query overload on the server side.
The genomic RNA of 'Severe_acute_respiratory_syndrome_coronavirus_2' has been downloaded to 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2' and has been named 'Severe_acute_respiratory_syndrome_coronavirus_2_rna_from_genomic_refseq.fna.gz' .


Starting Repeat Masker retrieval of 'Severe acute respiratory syndrome coronavirus 2' from refseq ...


|===========================================================================| 100%   53 MB
The download session seems to have timed out at the FTP site 'ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/858/895/GCF_009858895.2_ASM985889v3/GCF_009858895.2_ASM985889v3_rm.out.gz'. This could be due to an overload of queries to the databases. Please restart this function to continue the data retrieval process or wait for a while before restarting this function in case your IP address was logged due to an query overload on the server side.
The Repeat Masker output file of 'Severe_acute_respiratory_syndrome_coronavirus_2' has been downloaded to 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2' and has been named 'Severe_acute_respiratory_syndrome_coronavirus_2_rm_refseq.out.gz'.


Starting assembly quality stats retrieval of 'Severe acute respiratory syndrome coronavirus 2' from refseq ...


|===========================================================================| 100%   53 MB
|===========================================================================| 100%   53 MB
Genome assembly quality stats file download completed!
Checking md5 hash of file: sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt ...
The md5 hash of file 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2/Severe_acute_respiratory_syndrome_coronavirus_2_md5checksums.txt' matches!
The assembly statistics file of 'Severe_acute_respiratory_syndrome_coronavirus_2' has been downloaded to 'sars_cov_2/refseq/Severe_acute_respiratory_syndrome_coronavirus_2' and has been named 'Severe_acute_respiratory_syndrome_coronavirus_2_assembly_stats_refseq.txt'.
Collection retrieval finished successfully!</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># install.packages("readr")</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">sars_cov_<span class="dv">2</span>_info_all_strains &lt;-</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="st">  </span>readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span>(</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">    <span class="st">"sars_db/sars_cov_2_info_all_strains.csv"</span>,</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">    <span class="dt">col_types =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/cols.html">cols</a></span>(</a>
<a class="sourceLine" id="cb3-6" data-line-number="6">        <span class="dt">Accession =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">        <span class="dt">Release_Date =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_datetime.html">col_datetime</a></span>(<span class="dt">format =</span> <span class="st">""</span>),</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">        <span class="dt">Species =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">        <span class="dt">Genus =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">        <span class="dt">Family =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-11" data-line-number="11">        <span class="dt">Length =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_double</a></span>(),</a>
<a class="sourceLine" id="cb3-12" data-line-number="12">        <span class="dt">Nuc_Completeness =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-13" data-line-number="13">        <span class="dt">Genotype =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-14" data-line-number="14">        <span class="dt">Genome_Region =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-15" data-line-number="15">        <span class="dt">Segment =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-16" data-line-number="16">        <span class="dt">Authors =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-17" data-line-number="17">        <span class="dt">Publications =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_double</a></span>(),</a>
<a class="sourceLine" id="cb3-18" data-line-number="18">        <span class="dt">Geo_Location =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-19" data-line-number="19">        <span class="dt">Host =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-20" data-line-number="20">        <span class="dt">Isolation_Source =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-21" data-line-number="21">        <span class="dt">Collection_Date =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-22" data-line-number="22">        <span class="dt">BioSample =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>(),</a>
<a class="sourceLine" id="cb3-23" data-line-number="23">        <span class="dt">GenBank_Title =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/parse_atomic.html">col_character</a></span>()</a>
<a class="sourceLine" id="cb3-24" data-line-number="24">      ), <span class="dt">col_names =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb3-25" data-line-number="25">  )</a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(sars_cov_<span class="dv">2</span>_info_all_strains<span class="op">$</span>Species)</a></code></pre></div>
<pre><code>Severe acute respiratory syndrome-related coronavirus 
                                                 1147</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(sars_cov_<span class="dv">2</span>_info_all_strains<span class="op">$</span>Geo_Location)</a></code></pre></div>
<pre><code>  Australia     Belgium      Brazil       China     Finland       India        Iran 
         17           2          10         313          12          22          11 
      Italy       Japan    Malaysia       Nepal     Nigeria Philippines South Korea 
         14          11           3          10           1           4          10 
      Spain      Sweden      Taiwan    Thailand         USA    Viet Nam 
         29          10          30           2         553          20 </code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">biomartr<span class="op">::</span><span class="kw"><a href="https://docs.ropensci.org/biomartr/reference/getGroups.html">getGroups</a></span>(<span class="dt">kingdom =</span> <span class="st">"viral"</span>)</a></code></pre></div>
<pre><code>[1] "Ackermannviridae"    "Adenoviridae"        "Adomaviridae"       
  [4] "Alloherpesviridae"   "Alphaflexiviridae"   "Alphasatellitidae"  
  [7] "Alphatetraviridae"   "Alvernaviridae"      "Amalgaviridae"      
 [10] "Amnoonviridae"       "Ampullaviridae"      "Anelloviridae"      
 [13] "Arenaviridae"        "Arteriviridae"       "Ascoviridae"        
 [16] "Asfarviridae"        "Aspiviridae"         "Astroviridae"       
 [19] "Avsunviroidae"       "Bacilladnaviridae"   "Baculoviridae"      
 [22] "Barnaviridae"        "Benyviridae"         "Betaflexiviridae"   
 [25] "Bicaudaviridae"      "Bidnaviridae"        "Birnaviridae"       
 [28] "Botourmiaviridae"    "Bromoviridae"        "Caliciviridae"      
 [31] "Carmotetraviridae"   "Caulimoviridae"      "Chrysoviridae"      
 [34] "Chuviridae"          "Circoviridae"        "Clavaviridae"       
 [37] "Closteroviridae"     "Coronaviridae"       "Corticoviridae"     
 [40] "Cruliviridae"        "Cystoviridae"        "Deltaflexiviridae"  
 [43] "Dicistroviridae"     "Endornaviridae"      "Euroniviridae"      
 [46] "Filoviridae"         "Fimoviridae"         "Flaviviridae"       
 [49] "Fusariviridae"       "Fuselloviridae"      "Gammaflexiviridae"  
 [52] "Geminiviridae"       "Genomoviridae"       "Globuloviridae"     
 [55] "Guttaviridae"        "Hantaviridae"        "Hepadnaviridae"     
 [58] "Hepeviridae"         "Herelleviridae"      "Herpesviridae"      
 [61] "Hypoviridae"         "Hytrosaviridae"      "Iflaviridae"        
 [64] "Inoviridae"          "Iridoviridae"        "Kitaviridae"        
 [67] "Lavidaviridae"       "Leviviridae"         "Lipothrixviridae"   
 [70] "Luteoviridae"        "Malacoherpesviridae" "Marnaviridae"       
 [73] "Marseilleviridae"    "Matonaviridae"       "Medioniviridae"     
 [76] "Megabirnaviridae"    "Mesoniviridae"       "Metaviridae"        
 [79] "Microviridae"        "Mimiviridae"         "Mononiviridae"      
 [82] "Mymonaviridae"       "Myoviridae"          "Mypoviridae"        
 [85] "Nairoviridae"        "Nanoviridae"         "Narnaviridae"       
 [88] "Nimaviridae"         "Nodaviridae"         "Nudiviridae"        
 [91] "Nyamiviridae"        "Orthomyxoviridae"    "Other"              
 [94] "Ovaliviridae"        "Papillomaviridae"    "Paramyxoviridae"    
 [97] "Partitiviridae"      "Parvoviridae"        "Peribunyaviridae"   
[100] "Permutotetraviridae" "Phasmaviridae"       "Phenuiviridae"      
[103] "Phycodnaviridae"     "Picobirnaviridae"    "Picornaviridae"     
[106] "Pithoviridae"        "Plasmaviridae"       "Pleolipoviridae"    
[109] "Pneumoviridae"       "Podoviridae"         "Polycipiviridae"    
[112] "Polydnaviridae"      "Polyomaviridae"      "Portogloboviridae"  
[115] "Pospiviroidae"       "Potyviridae"         "Poxviridae"         
[118] "Qinviridae"          "Quadriviridae"       "Reoviridae"         
[121] "Retroviridae"        "Rhabdoviridae"       "Roniviridae"        
[124] "Rudiviridae"         "Secoviridae"         "Siphoviridae"       
[127] "Smacoviridae"        "Solemoviridae"       "Solinviviridae"     
[130] "Sphaerolipoviridae"  "Spiraviridae"        "Sunviridae"         
[133] "Tectiviridae"        "Tobaniviridae"       "Togaviridae"        
[136] "Tolecusatellitidae"  "Tombusviridae"       "Tospoviridae"       
[139] "Totiviridae"         "Tristromaviridae"    "Turriviridae"       
[142] "Tymoviridae"         "unclassified"        "Virgaviridae"       
[145] "Wupedeviridae"       "Xinmoviridae"        "Yueviridae"</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">biomartr<span class="op">::</span><span class="kw"><a href="https://docs.ropensci.org/biomartr/reference/listGenomes.html">listGenomes</a></span>(<span class="dt">db =</span> <span class="st">"refseq"</span>, <span class="dt">type =</span> <span class="st">"subgroup"</span>, <span class="dt">subset =</span> <span class="st">"Coronaviridae"</span>)</a></code></pre></div>
<pre><code>[1] "Human coronavirus 229E"                               
 [2] "Camel alphacoronavirus"                               
 [3] "Porcine epidemic diarrhea virus"                      
 [4] "Human coronavirus NL63"                               
 [5] "Human coronavirus HKU1"                               
 [6] "Swine enteric coronavirus"                            
 [7] "Rhinolophus bat coronavirus HKU2"                     
 [8] "Scotophilus bat coronavirus 512"                      
 [9] "Miniopterus bat coronavirus HKU8"                     
[10] "Rousettus bat coronavirus HKU9"                       
[11] "Tylonycteris bat coronavirus HKU4"                    
[12] "Pipistrellus bat coronavirus HKU5"                    
[13] "Severe acute respiratory syndrome-related coronavirus"
[14] "Severe acute respiratory syndrome coronavirus 2"      
[15] "Beluga whale coronavirus SW1"                         
[16] "Bat coronavirus BM48-31/BGR/2008"                     
[17] "Common moorhen coronavirus HKU21"                     
[18] "Magpie-robin coronavirus HKU18"                       
[19] "Night heron coronavirus HKU19"                        
[20] "Sparrow coronavirus HKU17"                            
[21] "White-eye coronavirus HKU16"                          
[22] "Wigeon coronavirus HKU20"                             
[23] "Rabbit coronavirus HKU14"                             
[24] "Rousettus bat coronavirus HKU10"                      
[25] "Ferret coronavirus"                                   
[26] "Middle East respiratory syndrome-related coronavirus" 
[27] "BtRf-AlphaCoV/YN2012"                                 
[28] "Bat coronavirus"                                      
[29] "Lucheng Rn rat coronavirus"                           
[30] "Wencheng Sm shrew coronavirus"                        
[31] "Bat coronavirus CDPHE15/USA/2006"                     
[32] "Coronavirus AcCoV-JC34"                               
[33] "Porcine coronavirus HKU15"                            
[34] "Betacoronavirus Erinaceus/VMC/DEU/2012"               
[35] "BtRf-AlphaCoV/HuB2013"                                
[36] "BtMr-AlphaCoV/SAX2011"                                
[37] "BtNv-AlphaCoV/SC2013"                                 
[38] "NL63-related bat coronavirus"                         
[39] "Betacoronavirus HKU24"                                
[40] "Bat Hp-betacoronavirus/Zhejiang2013"                  
[41] "Rousettus bat coronavirus"  </code></pre>
<p>all_data &lt;- biomartr::getKingdomAssemblySummary(db = “refseq”)</p>
<p>biomartr::is.genome.available(db = “refseq”, organism = “Severe acute respiratory syndrome coronavirus 2”, details = TRUE)</p>
<p>dplyr::filter(all_data, stringr::str_detect(organism_name, “Severe acute respiratory syndrome”))$organism_name [1] “Severe acute respiratory syndrome-related coronavirus” [2] “Severe acute respiratory syndrome coronavirus 2”</p>
</div>


  </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


