<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genomic Data Retrieval • biomartr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><meta property="og:title" content="Genomic Data Retrieval">
<meta property="og:description" content="Perform large scale genomic data retrieval and functional annotation retrieval. This package aims to provide users with a standardized
                way to automate genome, proteome, 'RNA', coding sequence ('CDS'), 'GFF', and metagenome
                retrieval from 'NCBI RefSeq', 'NCBI Genbank', 'ENSEMBL', 'ENSEMBLGENOMES',
                and 'UniProt' databases. Furthermore, an interface to the 'BioMart' database
                (Smedley et al. (2009) &lt;doi:10.1186/1471-2164-10-22&gt;) allows users to retrieve
                functional annotation for genomic loci. In addition, users can download entire databases such
                as 'NCBI RefSeq' (Pruitt et al. (2007) &lt;doi:10.1093/nar/gkl842&gt;), 'NCBI nr',
                'NCBI nt', 'NCBI Genbank' (Benson et al. (2013) &lt;doi:10.1093/nar/gks1195&gt;), etc. as
                well as 'ENSEMBL' and 'ENSEMBLGENOMES' with only one command.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">biomartr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.9.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/BioMart_Examples.html">Ensembl BioMart Examples</a>
    </li>
    <li>
      <a href="articles/Database_Retrieval.html">NCBI Database Retrieval</a>
    </li>
    <li>
      <a href="articles/Functional_Annotation.html">Functional Annotation with biomartr</a>
    </li>
    <li>
      <a href="articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a>
    </li>
    <li>
      <a href="articles/Sequence_Retrieval.html">Sequence Retrieval</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ropensci/biomartr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    
<div id="biomartr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#biomartr" class="anchor"></a>biomartr</h1></div>

<div id="genomic-data-retrieval-with-r" class="section level2">
<h2 class="hasAnchor">
<a href="#genomic-data-retrieval-with-r" class="anchor"></a>Genomic Data Retrieval with R</h2>
<div id="motivation" class="section level3">
<h3 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation:</h3>
<p>This package is born out of my own frustration to automate the genomic data retrieval process to create computationally reproducible scripts for large-scale genomics studies. Since I couldn’t find easy-to-use and fully reproducible software libraries I sat down and tried to implement a framework that would enable anyone to automate and standardize the genomic data retrieval process. I hope that this package is useful to others as well and that it helps to promote reproducible research in genomics studies.</p>
<p>I happily welcome anyone who wishes to contribute to this project :) Just drop me an email.</p>
</div>
<div id="short-package-description" class="section level3">
<h3 class="hasAnchor">
<a href="#short-package-description" class="anchor"></a>Short package description:</h3>
<p>The vastly growing number of sequenced genomes allows us to perform a new type of biological research. Using a comparative approach these genomes provide us with new insights on how biological information is encoded on the molecular level and how this information changes over evolutionary time.</p>
<p>The first step, however, of any genome based study is to retrieve genomes and their annotation from databases. To automate the retrieval process of this information on a meta-genomic scale, the <code>biomartr</code> package provides interface functions for genomic sequence retrieval and functional annotation retrieval. The major aim of <code>biomartr</code> is to facilitate computational reproducibility and large-scale handling of genomic data for (meta-)genomic analyses. In addition, <code>biomartr</code> aims to address the <code>genome version crisis</code>. With <code>biomartr</code> users can now control and be informed about the genome versions they retrieve automatically. Many large scale genomics studies lack this information and thus, reproducibility and data interpretation become nearly impossible when documentation of genome version information gets neglected.</p>
<p>In detail, <code>biomartr</code> automates genome, proteome, CDS, RNA, Repeats, GFF/GTF (annotation), genome assembly quality, and metagenome project data retrieval from the major biological databases such as</p>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/refseq/">NCBI RefSeq</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/genbank/">NCBI Genbank</a></li>
<li><a href="https://www.ensembl.org/index.html">ENSEMBL</a></li>
<li><a href="http://ensemblgenomes.org">ENSEMBLGENOMES</a></li>
<li><a href="http://www.uniprot.org">UniProt</a></li>
</ul>
<p>Furthermore, an interface to the <a href="www.ensembl.org/biomart">Ensembl Biomart</a> database allows users to retrieve functional annotation for genomic loci using a novel and organism centric search strategy. In addition, users can <a href="https://github.com/HajkD/biomartr/blob/master/vignettes/Database_Retrieval.Rmd">download entire databases</a> such as <code>NCBI RefSeq</code>, <code>NCBI nr</code>, <code>NCBI nt</code>, <code>NCBI Genbank</code>, etc. as well as <code>ENSEMBL</code> and <code>ENSEMBLGENOMES</code> with only one command.</p>
</div>
<div id="citation" class="section level3">
<h3 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h3>
<p><strong>I would be very greatful if you could cite the following paper in case <code>biomartr</code> was useful for your own research. I plan on vastly extending the biomartr functionality and usability in the next years. Many thanks in advance :)</strong></p>
<blockquote>
<p>Drost HG, Paszkowski J. <strong>Biomartr: genomic data retrieval with R</strong>. <em>Bioinformatics</em> (2017) 33(8): 1216-1217. <a href="https://academic.oup.com/bioinformatics/article/doi/10.1093/bioinformatics/btw821/2931816/Biomartr-genomic-data-retrieval-with-R">doi:10.1093/bioinformatics/btw821</a>.</p>
</blockquote>
</div>
<div id="similar-work" class="section level3">
<h3 class="hasAnchor">
<a href="#similar-work" class="anchor"></a>Similar Work</h3>
<p>The main difference between the <a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">BiomaRt</a> package and the <a href="https://ropensci.github.io/biomartr/">biomartr</a> package is that <code>biomartr</code> extends the <code>functional annotation retrieval</code> procedure of <code>BiomaRt</code> and <strong>in addition</strong> provides useful retrieval functions for genomes, proteomes, coding sequences, gff files, RNA sequences, Repeat Masker annotations files, and functions for the retrieval of entire databases such as <code>NCBI nr</code> etc.</p>
<p>Please consult the <a href="https://ropensci.github.io/biomartr/#tutorials">Tutorials section</a> for more details.</p>
<p><code>In the context of functional annotation retrieval</code> the <code>biomartr</code> package allows users to screen available marts using only the scientific name of an organism of interest instead of first searching for marts and datasets which support a particular organism of interest (which is required when using the <code>BiomaRt</code> package). Furthermore, <code>biomartr</code> allows you to search for particular topics when searching for attributes and filters. I am aware that the similar naming of the packages is unfortunate, but it arose due to historical reasons (please find a detailed explanation here: <a href="https://github.com/ropensci/biomartr/blob/master/FAQs.html" class="uri">https://github.com/ropensci/biomartr/blob/master/FAQs.md</a> and here #11).</p>
<p>I also dedicated <a href="https://ropensci.github.io/biomartr/articles/Functional_Annotation.html">an entire vignette to compare</a> the <code>BiomaRt</code> and <code>biomartr</code> package functionality in the context of <code>Functional Annotation</code> (where their functionality overlaps which comprises about only 20% of the overall functionality of the biomartr package).</p>
</div>
<div id="feedback" class="section level3">
<h3 class="hasAnchor">
<a href="#feedback" class="anchor"></a>Feedback</h3>
<blockquote>
<p><strong>I truly value your opinion and improvement suggestions. Hence, I would be extremely grateful if you could take this 1 minute and 3 question survey (<a href="https://goo.gl/forms/Qaoxxjb1EnNSLpM02" class="uri">https://goo.gl/forms/Qaoxxjb1EnNSLpM02</a>) so that I can learn how to improve <code>biomartr</code> in the best possible way. Many many thanks in advance.</strong></p>
</blockquote>
</div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install biomartr 0.8.0</span>
<span class="kw">install.packages</span>(<span class="st">"biomartr"</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<div id="collection-retrieval" class="section level3">
<h3 class="hasAnchor">
<a href="#collection-retrieval" class="anchor"></a>Collection Retrieval</h3>
<p>The automated retrieval of collections (= Genome, Proteome, CDS, RNA, GFF, Repeat Masker, AssemblyStats) will make sure that the genome file of an organism will match the CDS, proteome, RNA, GFF, etc file and was generated using the same genome assembly version. One aspect of why genomics studies fail in computational and biological reproducibility is that it is not clear whether CDS, proteome, RNA, GFF, etc files used in a proposed analysis were generated using the same genome assembly file denoting the same genome assembly version. To avoid this seemingly trivial mistake we encourage users to retrieve genome file collections using the <code>biomartr</code> function <code><a href="reference/getCollection.html">getCollection()</a></code> and attach the corresponding output as Supplementary Data to the respective genomics study to ensure computational and biological reproducibility.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download collection for Saccharomyces cerevisiae</span>
<span class="kw"><a href="reference/getCollection.html">getCollection</a></span>( <span class="dt">db =</span> <span class="st">"refseq"</span>, 
               <span class="dt">organism =</span> <span class="st">"Saccharomyces cerevisiae"</span>, 
               <span class="dt">path =</span> <span class="kw">file.path</span>(<span class="st">"refseq"</span>,<span class="st">"Collections"</span>))</code></pre></div>
<p>Internally, the <code><a href="reference/getCollection.html">getCollection()</a></code> function will now generate a folder named <code>refseq/Collection/Saccharomyces_cerevisiae</code> and will store all genome and annotation files for <code>Saccharomyces cerevisiae</code> in the same folder. In addition, the exact genoem and annotation version will be logged in the <code>doc</code> folder.</p>
<p>Internally, a text file named <code>doc_Saccharomyces_cerevisiae_db_refseq.txt</code> is generated. The information stored in this log file is structured as follows:</p>
<pre><code>File Name: Saccharomyces_cerevisiae_assembly_stats_refseq.txt
Organism Name: Saccharomyces_cerevisiae
Database: NCBI refseq
URL: ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_assembly_stats.txt
Download_Date: Wed Jun 27 15:21:51 2018
refseq_category: reference genome
assembly_accession: GCF_000146045.2
bioproject: PRJNA128
biosample: NA
taxid: 559292
infraspecific_name: strain=S288C
version_status: latest
release_type: Major
genome_rep: Full
seq_rel_date: 2014-12-17
submitter: Saccharomyces Genome Database</code></pre>
<p>In an ideal world this reference file could then be included as supplementary information in any life science publication that relies on genomic information so that reproducibility of experiments and analyses becomes achievable.</p>
</div>
<div id="genome-retrieval-of-hundreds-of-genomes-using-only-one-command" class="section level3">
<h3 class="hasAnchor">
<a href="#genome-retrieval-of-hundreds-of-genomes-using-only-one-command" class="anchor"></a>Genome retrieval of hundreds of genomes using only one command</h3>
<p>Download all mammalian vertebrate genomes from <code>NCBI RefSeq</code> via:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download all vertebrate genomes</span>
<span class="kw"><a href="reference/meta.retrieval.html">meta.retrieval</a></span>(<span class="dt">kingdom =</span> <span class="st">"vertebrate_mammalian"</span>, <span class="dt">db =</span> <span class="st">"refseq"</span>, <span class="dt">type =</span> <span class="st">"genome"</span>)</code></pre></div>
<p>All geneomes are stored in the folder named according to the kingdom. In this case <code>vertebrate_mammalian</code>. Alternatively, users can specify the <code>out.folder</code> argument to define a custom output folder path.</p>
</div>
<div id="platforms" class="section level3">
<h3 class="hasAnchor">
<a href="#platforms" class="anchor"></a>Platforms</h3>
<blockquote>
<p>Find <code>biomartr</code> also at <a href="https://omictools.com/biomartr-tool">OmicTools</a>.</p>
</blockquote>
</div>
<div id="frequently-asked-questions-faqs" class="section level3">
<h3 class="hasAnchor">
<a href="#frequently-asked-questions-faqs" class="anchor"></a>Frequently Asked Questions (FAQs)</h3>
<p>Please find <a href="FAQs.html">all FAQs here</a>.</p>
</div>
<div id="discussions-and-bug-reports" class="section level3">
<h3 class="hasAnchor">
<a href="#discussions-and-bug-reports" class="anchor"></a>Discussions and Bug Reports</h3>
<p>I would be very happy to learn more about potential improvements of the concepts and functions provided in this package.</p>
<p>Furthermore, in case you find some bugs or need additional (more flexible) functionality of parts of this package, please let me know:</p>
<p><a href="https://twitter.com/hajkdrost">twitter: HajkDrost</a> or <a href="hgd23@cam.ac.uk">email</a></p>
<p>For Bug Reports: Please send me an <a href="https://github.com/HajkD/biomartr/issues">issue</a>.</p>
</div>
</div>
<div id="tutorials" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorials" class="anchor"></a>Tutorials</h2>
<p>Getting Started with <code>biomartr</code>:</p>
<ul>
<li><a href="https://ropensci.github.io/biomartr/articles/Database_Retrieval.html">NCBI Database Retrieval</a></li>
<li><a href="https://ropensci.github.io/biomartr/articles/Sequence_Retrieval.html">Genomic Sequence Retrieval</a></li>
<li><a href="https://ropensci.github.io/biomartr/articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a></li>
<li><a href="https://ropensci.github.io/biomartr/articles/Functional_Annotation.html">Functional Annotation</a></li>
<li><a href="https://ropensci.github.io/biomartr/articles/BioMart_Examples.html">BioMart Examples</a></li>
</ul>
<p>Users can also read the tutorials within (<a href="http://www.rstudio.com/">RStudio</a>) :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># source the biomartr package</span>
<span class="kw">library</span>(biomartr)

<span class="co"># look for all tutorials (vignettes) available in the biomartr package</span>
<span class="co"># this will open your web browser</span>
<span class="kw">browseVignettes</span>(<span class="st">"biomartr"</span>)</code></pre></div>
</div>
<div id="news" class="section level2">
<h2 class="hasAnchor">
<a href="#news" class="anchor"></a>NEWS</h2>
<p>The current status of the package as well as a detailed history of the functionality of each version of <code>biomartr</code> can be found in the <a href="https://ropensci.github.io/biomartr/news/index.html">NEWS</a> section.</p>
</div>
<div id="install-developer-version" class="section level2">
<h2 class="hasAnchor">
<a href="#install-developer-version" class="anchor"></a>Install Developer Version</h2>
<p>Some bug fixes or new functionality will not be available on CRAN yet, but in the developer version here on GitHub. To download and install the most recent version of <code>biomartr</code> run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install the current version of biomartr on your system</span>
<span class="kw">source</span>(<span class="st">"http://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"ropensci/biomartr"</span>)</code></pre></div>
</div>
<div id="genomic-data-retrieval" class="section level2">
<h2 class="hasAnchor">
<a href="#genomic-data-retrieval" class="anchor"></a>Genomic Data Retrieval</h2>
<div id="meta-genome-retrieval" class="section level4">
<h4 class="hasAnchor">
<a href="#meta-genome-retrieval" class="anchor"></a>Meta-Genome Retrieval</h4>
<ul>
<li>
<code><a href="reference/meta.retrieval.html">meta.retrieval()</a></code> : Perform Meta-Genome Retieval from NCBI of species belonging to the same kingdom of life or to the same taxonomic subgroup</li>
<li>
<code><a href="reference/meta.retrieval.all.html">meta.retrieval.all()</a></code> : Perform Meta-Genome Retieval from NCBI of the entire kingdom of life</li>
<li>
<code><a href="reference/getMetaGenomes.html">getMetaGenomes()</a></code> : Retrieve metagenomes from NCBI Genbank</li>
<li>
<code><a href="reference/getMetaGenomeAnnotations.html">getMetaGenomeAnnotations()</a></code> : Retrieve annotation *.gff files for metagenomes from NCBI Genbank</li>
<li>
<code><a href="reference/listMetaGenomes.html">listMetaGenomes()</a></code> : List available metagenomes on NCBI Genbank</li>
<li>
<code><a href="reference/getMetaGenomeSummary.html">getMetaGenomeSummary()</a></code> : Helper function to retrieve the assembly_summary.txt file from NCBI genbank metagenomes</li>
<li>
<code><a href="reference/clean.retrieval.html">clean.retrieval()</a></code>: Format meta.retrieval output</li>
</ul>
</div>
<div id="genome-retrieval" class="section level4">
<h4 class="hasAnchor">
<a href="#genome-retrieval" class="anchor"></a>Genome Retrieval</h4>
<ul>
<li>
<code><a href="reference/listGenomes.html">listGenomes()</a></code> : List all genomes available on NCBI and ENSEMBL servers</li>
<li>
<code><a href="reference/listKingdoms.html">listKingdoms()</a></code> : list the number of available species per kingdom of life on NCBI and ENSEMBL servers</li>
<li>
<code><a href="reference/listGroups.html">listGroups()</a></code> : list the number of available species per group on NCBI and ENSEMBL servers</li>
<li>
<code><a href="reference/getKingdoms.html">getKingdoms()</a></code> : Retrieve available kingdoms of life</li>
<li>
<code><a href="reference/getGroups.html">getGroups()</a></code> : Retrieve available groups for a kingdom of life</li>
<li>
<code><a href="reference/is.genome.available.html">is.genome.available()</a></code> : Check Genome Availability NCBI and ENSEMBL servers</li>
<li>
<code><a href="reference/getCollection.html">getCollection()</a></code> : Retrieve a Collection: Genome, Proteome, CDS, RNA, GFF, Repeat Masker, AssemblyStats</li>
<li>
<code><a href="reference/getGenome.html">getGenome()</a></code> : Download a specific genome stored on NCBI and ENSEMBL servers</li>
<li>
<code><a href="reference/getProteome.html">getProteome()</a></code> : Download a specific proteome stored on NCBI and ENSEMBL servers</li>
<li>
<code><a href="reference/getCDS.html">getCDS()</a></code> : Download a specific CDS file (genome) stored on NCBI and ENSEMBL servers</li>
<li>
<code><a href="reference/getRNA.html">getRNA()</a></code> : Download a specific RNA file stored on NCBI and ENSEMBL servers</li>
<li>
<code><a href="reference/getGFF.html">getGFF()</a></code> : Genome Annotation Retrieval from NCBI (<code>*.gff</code>) and ENSEMBL (<code>*.gff3</code>) servers</li>
<li>
<code><a href="reference/getGTF.html">getGTF()</a></code> : Genome Annotation Retrieval (<code>*.gtf</code>) from ENSEMBL servers</li>
<li>
<code>getRepeatMasker() :</code> Repeat Masker TE Annotation Retrieval</li>
<li>
<code><a href="reference/getAssemblyStats.html">getAssemblyStats()</a></code> : Genome Assembly Stats Retrieval from NCBI</li>
<li>
<code><a href="reference/getKingdomAssemblySummary.html">getKingdomAssemblySummary()</a></code> : Helper function to retrieve the assembly_summary.txt files from NCBI for all kingdoms</li>
<li>
<code><a href="reference/getMetaGenomeSummary.html">getMetaGenomeSummary()</a></code> : Helper function to retrieve the assembly_summary.txt files from NCBI genbank metagenomes</li>
<li>
<code><a href="reference/getSummaryFile.html">getSummaryFile()</a></code> : Helper function to retrieve the assembly_summary.txt file from NCBI for a specific kingdom</li>
<li>
<code><a href="reference/getENSEMBLInfo.html">getENSEMBLInfo()</a></code> : Retrieve ENSEMBL info file</li>
<li>
<code><a href="reference/getGENOMEREPORT.html">getGENOMEREPORT()</a></code> : Retrieve GENOME_REPORTS file from NCBI</li>
</ul>
</div>
<div id="import-downloaded-files" class="section level4">
<h4 class="hasAnchor">
<a href="#import-downloaded-files" class="anchor"></a>Import Downloaded Files</h4>
<ul>
<li>
<code><a href="reference/read_genome.html">read_genome()</a></code> : Import genomes as Biostrings or data.table object</li>
<li>
<code><a href="reference/read_proteome.html">read_proteome()</a></code> : Import proteome as Biostrings or data.table object</li>
<li>
<code><a href="reference/read_cds.html">read_cds()</a></code> : Import CDS as Biostrings or data.table object</li>
<li>
<code><a href="reference/read_gff.html">read_gff()</a></code> : Import GFF file</li>
<li>
<code><a href="reference/read_rna.html">read_rna()</a></code> : Import RNA file</li>
<li>
<code><a href="reference/read_rm.html">read_rm()</a></code> : Import Repeat Masker output file</li>
<li>
<code><a href="reference/read_assemblystats.html">read_assemblystats()</a></code> : Import Genome Assembly Stats File</li>
</ul>
</div>
<div id="database-retrieval" class="section level4">
<h4 class="hasAnchor">
<a href="#database-retrieval" class="anchor"></a>Database Retrieval</h4>
<ul>
<li>
<code><a href="reference/listDatabases.html">listNCBIDatabases()</a></code> : Retrieve a List of Available NCBI Databases for Download</li>
<li>
<code><a href="reference/download.database.html">download.database()</a></code> : Download a NCBI database to your local hard drive</li>
<li>
<code><a href="reference/download.database.all.html">download.database.all()</a></code> : Download a complete NCBI Database such as e.g. <code>NCBI nr</code> to your local hard drive</li>
</ul>
</div>
<div id="biomart-queries" class="section level3">
<h3 class="hasAnchor">
<a href="#biomart-queries" class="anchor"></a>BioMart Queries</h3>
<ul>
<li>
<code><a href="reference/biomart.html">biomart()</a></code> : Main function to query the BioMart database</li>
<li>
<code><a href="reference/getMarts.html">getMarts()</a></code> : Retrieve All Available BioMart Databases</li>
<li>
<code><a href="reference/getDatasets.html">getDatasets()</a></code> : Retrieve All Available Datasets for a BioMart Database</li>
<li>
<code><a href="reference/getAttributes.html">getAttributes()</a></code> : Retrieve All Available Attributes for a Specific Dataset</li>
<li>
<code><a href="reference/getFilters.html">getFilters()</a></code> : Retrieve All Available Filters for a Specific Dataset</li>
<li>
<code><a href="reference/organismBM.html">organismBM()</a></code> : Function for organism specific retrieval of available BioMart marts and datasets</li>
<li>
<code><a href="reference/organismAttributes.html">organismAttributes()</a></code> : Function for organism specific retrieval of available BioMart attributes</li>
<li>
<code><a href="reference/organismFilters.html">organismFilters()</a></code> : Function for organism specific retrieval of available BioMart filters</li>
</ul>
</div>
<div id="performing-gene-ontology-queries" class="section level3">
<h3 class="hasAnchor">
<a href="#performing-gene-ontology-queries" class="anchor"></a>Performing Gene Ontology queries</h3>
<div id="gene-ontology" class="section level4">
<h4 class="hasAnchor">
<a href="#gene-ontology" class="anchor"></a>Gene Ontology</h4>
<ul>
<li>
<code><a href="reference/getGO.html">getGO()</a></code> : Function to retrieve GO terms for a given set of genes</li>
</ul>
</div>
</div>
</div>
<div id="download-developer-version-on-windows-systems" class="section level2">
<h2 class="hasAnchor">
<a href="#download-developer-version-on-windows-systems" class="anchor"></a>Download Developer Version On Windows Systems</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># On Windows, this won't work - see ?build_github_devtools</span>
<span class="kw">install_github</span>(<span class="st">"HajkD/biomartr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)

<span class="co"># When working with Windows, first you need to install the</span>
<span class="co"># R package: rtools -&gt; install.packages("rtools")</span>

<span class="co"># Afterwards you can install devtools -&gt; install.packages("devtools")</span>
<span class="co"># and then you can run:</span>

devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"HajkD/biomartr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)

<span class="co"># and then call it from the library</span>
<span class="kw">library</span>(<span class="st">"biomartr"</span>, <span class="dt">lib.loc =</span> <span class="st">"C:/Program Files/R/R-3.1.1/library"</span>)</code></pre></div>
<div id="troubleshooting-on-windows-machines" class="section level3">
<h3 class="hasAnchor">
<a href="#troubleshooting-on-windows-machines" class="anchor"></a>Troubleshooting on Windows Machines</h3>
<ul>
<li>Install <code>biomartr</code> on a Win 8 laptop: <a href="https://github.com/HajkD/orthologr/issues/1">solution</a> ( Thanks to Andres Romanowski )</li>
</ul>
</div>
</div>
</div>
<div id="code-of-conduct" class="section level1">
<h1 class="hasAnchor">
<a href="#code-of-conduct" class="anchor"></a>Code of conduct</h1>
<p>Please note that this project is released with a <a href="CONDUCT.html">Contributor Code of Conduct</a>. By participating in this project you agree to abide by its terms.</p>
</div>

  </div>

  <div class="col-md-3" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=biomartr">https://​cloud.r-project.org/​package=biomartr</a>
</li>
<li>Browse source code at <br><a href="https://github.com/ropensci/biomartr">https://​github.com/​ropensci/​biomartr</a>
</li>
<li>Report a bug at <br><a href="https://github.com/ropensci/biomartr/issues">https://​github.com/​ropensci/​biomartr/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing biomartr</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Hajk-Georg Drost <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/http://orcid.org/0000-0002-1567-306X" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" height="16"></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.rpackages.io/package/biomartr"><img src="https://www.rpackages.io/badge/biomartr.svg" alt="rpackages.io rank"></a></li>
<li><a href="https://github.com/ropensci/onboarding/issues/93"><img src="https://badges.ropensci.org/93_status.svg"></a></li>
<li><a href="https://github.com/metacran/cranlogs.app"><img src="http://cranlogs.r-pkg.org/badges/biomartr" alt="rstudio mirror downloads"></a></li>
<li><a href="https://github.com/metacran/cranlogs.app"><img src="http://cranlogs.r-pkg.org/badges/grand-total/biomartr" alt="rstudio mirror downloads"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
