<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Functional Annotation with biomartr • biomartr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Functional Annotation with biomartr">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biomartr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.9.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BioMart_Examples.html">Ensembl BioMart Examples</a>
    </li>
    <li>
      <a href="../articles/Database_Retrieval.html">NCBI Database Retrieval</a>
    </li>
    <li>
      <a href="../articles/Functional_Annotation.html">Functional Annotation with biomartr</a>
    </li>
    <li>
      <a href="../articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a>
    </li>
    <li>
      <a href="../articles/Sequence_Retrieval.html">Sequence Retrieval</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ropensci/biomartr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Functional Annotation with biomartr</h1>
            
            <h4 class="date">2018-06-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ropensci/biomartr/blob/master/vignettes/Functional_Annotation.Rmd"><code>vignettes/Functional_Annotation.Rmd</code></a></small>
      <div class="hidden name"><code>Functional_Annotation.Rmd</code></div>

    </div>

    
    
<div id="functional-annotation-retrieval-from-ensembl-biomart" class="section level2">
<h2 class="hasAnchor">
<a href="#functional-annotation-retrieval-from-ensembl-biomart" class="anchor"></a>Functional Annotation Retrieval from <code>Ensembl Biomart</code>
</h2>
<p>The <a href="http://ensemblgenomes.org/info/access/biomart">Ensembl Biomart</a> database enables users to retrieve a vast diversity of annotation data for specific organisms. Initially, Steffen Durinck and Wolfgang Huber provided a powerful interface between the R language and <a href="http://ensemblgenomes.org/info/access/biomart">Ensembl Biomart</a> by implementing the R package <a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a>.</p>
<p>The purpose of the <code>biomaRt</code> package was to mimic the ENSEMBL BioMart database structure to construct queries that can be sent to the Application Programming Interface (API) of BioMart. Although, this procedure was very useful in the past, it seems not intuitive from an organism centric point of view. Usually, users wish to download functional annotation for a particular organism of interest. However, the BioMart and thus the <code>biomaRt</code> package require that users already know in which <code>mart</code> and <code>dataset</code> the organism of interest will be found which requires significant efforts of searching and screening. In addition, once the <code>mart</code> and <code>dataset</code> of a particular organism of interest were found and specified the user must again learn which <code>attribute</code> has to be specified to retrieve the functional annotation information of interest.</p>
<p>The new functionality implemented in the <code>biomartr</code> package aims to overcome this search bottleneck by extending the functionality of the <a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> package. The new <code>biomartr</code> package introduces a more organism cantered annotation retrieval concept which does not require to screen for <code>marts</code>, <code>datasets</code>, and <code>attributes</code> beforehand. With <code>biomartr</code> users only need to specify the <code>scientific name</code> of the organism of interest to then retrieve available <code>marts</code>, <code>datasets</code>, and <code>attributes</code> for the corresponding organism of interest.</p>
<p>This paradigm shift enables users to quickly construct queries to the BioMart database without having to learn the particular database structure and organization of BioMart.</p>
<p>The following sections will introduce users to the functionality and data retrieval precedures of <code>biomartr</code> and will show how <code>biomartr</code> extends the functionality of the initial <a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> package.</p>
<div id="the-old-biomart-query-methodology" class="section level3">
<h3 class="hasAnchor">
<a href="#the-old-biomart-query-methodology" class="anchor"></a>The old <code>biomaRt</code> query methodology</h3>
<p>The best way to get started with the <em>old</em> methodology presented by the established <a href="http://www.bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> package is to understand the workflow of its data retrieval process. The query logic of the <code>biomaRt</code> package derives from the database organization of <a href="http://ensemblgenomes.org/info/access/biomart">Ensembl Biomart</a> which stores a vast diversity of annotation data for specific organisms. In detail, the <a href="http://ensemblgenomes.org/info/access/biomart">Ensembl Biomart</a> database is organized into so called:<br><code>marts</code>, <code>datasets</code>, and <code>attributes</code>. <code>Marts</code> denote a higher level category of functional annotation such as <code>SNP</code> (e.g. for functional annotation of particular single nucleotide polymorphisms (SNPs)) or <code>FUNCGEN</code> (e.g. for functional annotation of regulatory regions or relationsships of genes). <code>Datasets</code> denote the particular species of interest for which functional annotation is available <strong>within</strong> this specific <code>mart</code>. It can happen that <code>datasets</code> (= particular species of interest) are available in one <code>mart</code> (= higher category of functional annotation) but not in an other <code>mart</code>. For the actual retrieval of functional annotation information users must then specify the <code>type</code> of functional annotation information they wish to retrieve. These <code>types</code> are called <code>attributes</code> in the <code>biomaRt</code> notation.</p>
<p>Hence, when users wish to retrieve information for a specific organism of interest, they first need to specify a particular <code>mart</code> and <code>dataset</code> in which the information of the corresponding organism of interest can be found. Subsequently they can specify the <code>attributes</code> argument to retrieve a particular type of functional annotation (e.g. Gene Ontology terms).</p>
<p>The following section shall illustrate how <code>marts</code>, <code>datasets</code>, and <code>attributes</code> could be explored using <code>biomaRt</code> before the <code>biomartr</code> package existed.</p>
<p>The availability of <code>marts</code>, <code>datasets</code>, and <code>attributes</code> can be checked by the following functions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install the biomaRt package       </span>
<span class="co"># source("http://bioconductor.org/biocLite.R")      </span>
<span class="co"># biocLite("biomaRt")       </span>
<span class="co"># load biomaRt      </span>
<span class="kw">library</span>(biomaRt)        
<span class="co"># look at top 10 databases      </span>
<span class="kw">head</span>(biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/listMarts">listMarts</a></span>(<span class="dt">host =</span> <span class="st">"www.ensembl.org"</span>), <span class="dv">10</span>)      </code></pre></div>
<pre><code>#&gt;                biomart               version
#&gt; 1 ENSEMBL_MART_ENSEMBL      Ensembl Genes 92
#&gt; 2   ENSEMBL_MART_MOUSE      Mouse strains 92
#&gt; 3     ENSEMBL_MART_SNP  Ensembl Variation 92
#&gt; 4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 92</code></pre>
<p>Users will observe that several <code>marts</code> providing annotation for specific classes of organisms or groups of organisms are available.</p>
<p>For our example, we will choose the <code>hsapiens_gene_ensembl</code> <code>mart</code> and list all available datasets that are element of this <code>mart</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/listDatasets">listDatasets</a></span>(biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ENSEMBL_MART_ENSEMBL"</span>, <span class="dt">host =</span> <span class="st">"www.ensembl.org"</span>)), <span class="dv">10</span>)     </code></pre></div>
<pre><code>#&gt;                        dataset                                  description            version
#&gt; 1   acarolinensis_gene_ensembl               Anole lizard genes (AnoCar2.0)          AnoCar2.0
#&gt; 2    amelanoleuca_gene_ensembl                        Panda genes (ailMel1)            ailMel1
#&gt; 3      amexicanus_gene_ensembl                  Cave fish genes (AstMex102)          AstMex102
#&gt; 4      anancymaae_gene_ensembl           Ma's night monkey genes (Anan_2.0)           Anan_2.0
#&gt; 5  aplatyrhynchos_gene_ensembl                    Duck genes (BGI_duck_1.0)       BGI_duck_1.0
#&gt; 6         btaurus_gene_ensembl                           Cow genes (UMD3.1)             UMD3.1
#&gt; 7         caperea_gene_ensembl        Brazilian guinea pig genes (CavAp1.0)           CavAp1.0
#&gt; 8           catys_gene_ensembl              Sooty mangabey genes (Caty_1.0)           Caty_1.0
#&gt; 9      ccapucinus_gene_ensembl          Capuchin genes (Cebus_imitator-1.0) Cebus_imitator-1.0
#&gt; 10     cchok1gshd_gene_ensembl Chinese hamster CHOK1GS genes (CHOK1GS_HDv1)       CHOK1GS_HDv1</code></pre>
<p>The <code><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart()</a></code> function is a wrapper function provided by <code>biomaRt</code> to connect a selected BioMart database (<code>mart</code>) with a corresponding dataset stored within this <code>mart</code>.<br>
We select dataset <code>hsapiens_gene_ensembl</code> and now check for available attributes (annotation data) that can be accessed for <code>Homo sapiens</code> genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/listAttributes">listAttributes</a></span>(biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useDataset">useDataset</a></span>(
                                         <span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>,         
                                         <span class="dt">mart    =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ENSEMBL_MART_ENSEMBL"</span>,      
                                         <span class="dt">host    =</span> <span class="st">"www.ensembl.org"</span>))), <span class="dv">10</span>)        </code></pre></div>
<p>Please note the nested structure of this attribute query. For an attribute query procedure an additional wrapper function named <code><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useDataset">useDataset()</a></code> is needed in which <code><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart()</a></code> and a corresponding dataset needs to be specified. The result is a table storing the name of available attributes for<br><em>Homo sapiens</em> as well as a short description.</p>
<p>Furthermore, users can retrieve all filters for <em>Homo sapiens</em> that can be specified by the actual BioMart query process.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">head</span>(biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/listFilters">listFilters</a></span>(biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useDataset">useDataset</a></span>(<span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>,       
                                               <span class="dt">mart    =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ENSEMBL_MART_ENSEMBL"</span>,        
                                               <span class="dt">host    =</span> <span class="st">"www.ensembl.org"</span>))), <span class="dv">10</span>)      </code></pre></div>
<p>After accumulating all this information, it is now possible to perform an actual BioMart query by using the <code><a href="http://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM()</a></code> function.</p>
<p>In this example we will retrieve attributes: <code>start_position</code>,<code>end_position</code> and <code>description</code><br>
for the <em>Homo sapiens</em> gene <code>"GUCA2A"</code>.</p>
<p>Since the input genes are <code>ensembl gene ids</code>, we need to specify the <code>filters</code> argument <code>filters = "hgnc_symbol"</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="co"># 1) select a mart and data set        </span>
 mart &lt;-<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useDataset">useDataset</a></span>(<span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>,         
                    <span class="dt">mart    =</span> <span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>(<span class="st">"ENSEMBL_MART_ENSEMBL"</span>,       
                    <span class="dt">host    =</span> <span class="st">"www.ensembl.org"</span>))       
        
 <span class="co"># 2) run a biomart query using the getBM() function        </span>
 <span class="co"># and specify the attributes and filter arguments      </span>
 geneSet &lt;-<span class="st"> "GUCA2A"</span>        
        
 resultTable &lt;-<span class="st"> </span>biomaRt<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM</a></span>(<span class="dt">attributes =</span> <span class="kw">c</span>(<span class="st">"start_position"</span>,<span class="st">"end_position"</span>,<span class="st">"description"</span>),       
                      <span class="dt">filters    =</span> <span class="st">"hgnc_symbol"</span>,       
                      <span class="dt">values     =</span> geneSet,         
                      <span class="dt">mart       =</span> mart)        
        
 resultTable        </code></pre></div>
<p>When using <code><a href="http://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM()</a></code> users can pass all attributes retrieved by <code><a href="http://www.rdocumentation.org/packages/biomaRt/topics/listAttributes">listAttributes()</a></code> to the <code>attributes</code> argument of the <code><a href="http://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM()</a></code> function.</p>
</div>
</div>
<div id="extending-biomart-using-the-new-query-system-of-the-biomartr-package" class="section level2">
<h2 class="hasAnchor">
<a href="#extending-biomart-using-the-new-query-system-of-the-biomartr-package" class="anchor"></a>Extending <code>biomaRt</code> using the new query system of the <code>biomartr</code> package</h2>
<div id="getting-started-with-biomartr" class="section level3">
<h3 class="hasAnchor">
<a href="#getting-started-with-biomartr" class="anchor"></a>Getting Started with <code>biomartr</code>
</h3>
<p>This query methodology provided by <code>Ensembl Biomart</code> and the <code>biomaRt</code> package is a very well defined approach for accurate annotation retrieval. Nevertheless, when learning this query methodology it (subjectively) seems non-intuitive from the user perspective. Therefore, the <code>biomartr</code> package provides another query methodology that aims to be more organism centric.</p>
<p>Taken together, the following workflow allows users to perform fast BioMart queries for attributes using the <code><a href="../reference/biomart.html">biomart()</a></code> function implemented in this <code>biomartr</code> package:</p>
<ol style="list-style-type: decimal">
<li><p>get attributes, datasets, and marts via : <code><a href="../reference/organismAttributes.html">organismAttributes()</a></code></p></li>
<li><p>choose available biological features (filters) via: <code><a href="../reference/organismFilters.html">organismFilters()</a></code></p></li>
<li><p>specify a set of query genes: e.g. retrieved with <code><a href="../reference/getGenome.html">getGenome()</a></code>, <code><a href="../reference/getProteome.html">getProteome()</a></code> or <code><a href="../reference/getCDS.html">getCDS()</a></code></p></li>
<li><p>specify all arguments of the <code><a href="../reference/biomart.html">biomart()</a></code> function using steps 1) - 3) and perform a BioMart query</p></li>
</ol>
<p><strong>Note that dataset names change very frequently due to the update of dataset versions. So in case some query functions do not work properly, users should check with <code><a href="../reference/organismAttributes.html">organismAttributes(update = TRUE)</a></code> whether or not their dataset name has been changed. For example, <code><a href="../reference/organismAttributes.html">organismAttributes("Homo sapiens", topic = "id", update = TRUE)</a></code> might reveal that the dataset <code>ENSEMBL_MART_ENSEMBL</code> has changed.</strong></p>
</div>
</div>
<div id="retrieve-marts-datasets-attributes-and-filters-with-biomartr" class="section level2">
<h2 class="hasAnchor">
<a href="#retrieve-marts-datasets-attributes-and-filters-with-biomartr" class="anchor"></a>Retrieve marts, datasets, attributes, and filters with biomartr</h2>
<div id="retrieve-available-marts" class="section level3">
<h3 class="hasAnchor">
<a href="#retrieve-available-marts" class="anchor"></a>Retrieve Available Marts</h3>
<p>The <code><a href="../reference/getMarts.html">getMarts()</a></code> function allows users to list all available databases that can be accessed through BioMart interfaces.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the biomartr package</span>
<span class="kw">library</span>(biomartr)

<span class="co"># list all available databases</span>
biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getMarts">getMarts</a></span>()</code></pre></div>
</div>
<div id="retrieve-available-datasets-from-a-specific-mart" class="section level3">
<h3 class="hasAnchor">
<a href="#retrieve-available-datasets-from-a-specific-mart" class="anchor"></a>Retrieve Available Datasets from a Specific Mart</h3>
<p>Now users can select a specific database to list all available datasets that can be accessed through this database. In this example we choose the <code>ENSEMBL_MART_ENSEMBL</code> database.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getDatasets">getDatasets</a></span>(<span class="dt">mart =</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>) , <span class="dv">5</span>)</code></pre></div>
<p>Now you can select the dataset <code>hsapiens_gene_ensembl</code> and list all available attributes that can be retrieved from this dataset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tail</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getDatasets">getDatasets</a></span>(<span class="dt">mart =</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>) , <span class="dv">38</span>)</code></pre></div>
</div>
<div id="retrieve-available-attributes-from-a-specific-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#retrieve-available-attributes-from-a-specific-dataset" class="anchor"></a>Retrieve Available Attributes from a Specific Dataset</h3>
<p>Now that you have selected a database (<code>hsapiens_gene_ensembl</code>) and a dataset (<code>hsapiens_gene_ensembl</code>), users can list all available attributes for this dataset using the <code><a href="../reference/getAttributes.html">getAttributes()</a></code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># list all available attributes for dataset: hsapiens_gene_ensembl</span>
<span class="kw">head</span>( biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getAttributes">getAttributes</a></span>(<span class="dt">mart    =</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, 
                              <span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>), <span class="dv">10</span> )</code></pre></div>
</div>
<div id="retrieve-available-filters-from-a-specific-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#retrieve-available-filters-from-a-specific-dataset" class="anchor"></a>Retrieve Available Filters from a Specific Dataset</h3>
<p>Finally, the <code><a href="../reference/getFilters.html">getFilters()</a></code> function allows users to list available filters for a specific dataset that can be used for a <code><a href="../reference/biomart.html">biomart()</a></code> query.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># list all available filters for dataset: hsapiens_gene_ensembl</span>
<span class="kw">head</span>( biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getFilters">getFilters</a></span>(<span class="dt">mart    =</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, 
                           <span class="dt">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>), <span class="dv">10</span> )</code></pre></div>
</div>
</div>
<div id="organism-specific-retrieval-of-information" class="section level2">
<h2 class="hasAnchor">
<a href="#organism-specific-retrieval-of-information" class="anchor"></a>Organism Specific Retrieval of Information</h2>
<p>In most use cases, users will work with a single or a set of model organisms. In this process they will mostly be interested in specific annotations for this particular model organism. The <code><a href="../reference/organismBM.html">organismBM()</a></code> function addresses this issue and provides users with an organism centric query to <code>marts</code> and <code>datasets</code> which are available for a particular organism of interest.</p>
<p><strong>Note</strong> that when running the following functions for the first time, the data retrieval procedure will take some time, due to the remote access to BioMart. The corresponding result is then saved in a <code>*.txt</code> file named <code>_biomart/listDatasets.txt</code> within the <code>tempdir()</code> folder, allowing subsequent queries to be performed much faster. The <code>tempdir()</code> folder, however, will be deleted after a new R session was established. In this case the inital call of the subsequent functions again will take time to retrieve all organism specific data from the BioMart database.</p>
<p>This concept of locally storing all organism specific database linking information available in BioMart into an internal file allows users to significantly speed up subsequent retrieval queries for that particular organism.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># retrieving all available datasets and biomart connections for</span>
<span class="co"># a specific query organism (scientific name)</span>
biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismBM">organismBM</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>)</code></pre></div>
<p>The result is a table storing all <code>marts</code> and <code>datasets</code> from which annotations can be retrieved for <em>Homo sapiens</em>. Furthermore, a short description as well as the version of the dataset being accessed (very useful for publications) is returned.</p>
<p>Users will observe that 3 different <code>marts</code> provide 6 different <code>datasets</code> storing annotation information for <em>Homo sapiens</em>.</p>
<p><strong>Please note, however, that scientific names of organisms must be written correctly! For ex. “Homo Sapiens” will be treated differently (not recognized) than “Homo sapiens” (recognized).</strong></p>
<p>Similar to the <code>biomaRt</code> package query methodology, users need to specify <code>attributes</code> and <code>filters</code> to be able to perform accurate BioMart queries. Here the functions <code><a href="../reference/organismAttributes.html">organismAttributes()</a></code> and <code><a href="../reference/organismFilters.html">organismFilters()</a></code> provide useful and intuitive concepts to obtain this information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># return available attributes for "Homo sapiens"</span>
<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismAttributes">organismAttributes</a></span>(<span class="st">"Homo sapiens"</span>), <span class="dv">20</span>)</code></pre></div>
<p>Users will observe that the <code><a href="../reference/organismAttributes.html">organismAttributes()</a></code> function returns a data.frame storing attribute names, datasets, and marts which are available for <code>Homo sapiens</code>. After the ENSEMBL release 87 the <code>ENSEMBL_MART_SEQUENCE</code> service provided by Ensembl does not work properly and thus the <code><a href="../reference/organismAttributes.html">organismAttributes()</a></code> function prints out warning messages to make the user aware when certain marts provided bt Ensembl do not work properly, yet.</p>
<p>An additional feature provided by <code><a href="../reference/organismAttributes.html">organismAttributes()</a></code> is the <code>topic</code> argument. The <code>topic</code> argument allows users to to search for specific attributes, topics, or categories for faster filtering.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># search for attribute topic "id"</span>
<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismAttributes">organismAttributes</a></span>(<span class="st">"Homo sapiens"</span>, <span class="dt">topic =</span> <span class="st">"id"</span>), <span class="dv">20</span>)</code></pre></div>
<p>Now, all <code>attribute names</code> having <code>id</code> as part of their <code>name</code> are being returned.</p>
<p>Another example is <code>topic = "homolog"</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># search for attribute topic "homolog"</span>
<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismAttributes">organismAttributes</a></span>(<span class="st">"Homo sapiens"</span>, <span class="dt">topic =</span> <span class="st">"homolog"</span>), <span class="dv">20</span>)</code></pre></div>
<p>Or <code>topic = "dn"</code> and <code>topic = "ds"</code> for <code>dn</code> and <code>ds</code> value retrieval.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># search for attribute topic "dn"</span>
<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismAttributes">organismAttributes</a></span>(<span class="st">"Homo sapiens"</span>, <span class="dt">topic =</span> <span class="st">"dn"</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># search for attribute topic "ds"</span>
<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismAttributes">organismAttributes</a></span>(<span class="st">"Homo sapiens"</span>, <span class="dt">topic =</span> <span class="st">"ds"</span>))</code></pre></div>
<p>Analogous to the <code><a href="../reference/organismAttributes.html">organismAttributes()</a></code> function, the <code><a href="../reference/organismFilters.html">organismFilters()</a></code> function returns all filters that are available for a query organism of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># return available filters for "Homo sapiens"</span>
<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismFilters">organismFilters</a></span>(<span class="st">"Homo sapiens"</span>), <span class="dv">20</span>)</code></pre></div>
<p>The <code><a href="../reference/organismFilters.html">organismFilters()</a></code> function also allows users to search for filters that correspond to a specific topic or category.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># search for filter topic "id"</span>
<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismFilters">organismFilters</a></span>(<span class="st">"Homo sapiens"</span>, <span class="dt">topic =</span> <span class="st">"id"</span>), <span class="dv">20</span>)</code></pre></div>
</div>
<div id="construct-biomart-queries-with-biomartr" class="section level2">
<h2 class="hasAnchor">
<a href="#construct-biomart-queries-with-biomartr" class="anchor"></a>Construct BioMart queries with <code>biomartr</code>
</h2>
<p>The short introduction to the functionality of <code><a href="../reference/organismBM.html">organismBM()</a></code>, <code><a href="../reference/organismAttributes.html">organismAttributes()</a></code>, and <code><a href="../reference/organismFilters.html">organismFilters()</a></code> will allow users to perform BioMart queries in a very intuitive organism centric way. The main function to perform BioMart queries is <code><a href="../reference/biomart.html">biomart()</a></code>.</p>
<p>For the following examples we will assume that we are interested in the annotation of specific genes from the <em>Homo sapiens</em> proteome. We want to map the corresponding refseq gene id to a set of other gene ids used in other databases. For this purpose, first we need consult the <code><a href="../reference/organismAttributes.html">organismAttributes()</a></code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)

<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/organismAttributes">organismAttributes</a></span>(<span class="st">"Homo sapiens"</span>, <span class="dt">topic =</span> <span class="st">"id"</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># retrieve the proteome of Homo sapiens from refseq</span>
file_path &lt;-<span class="st"> </span>biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getProteome">getProteome</a></span>( <span class="dt">db       =</span> <span class="st">"refseq"</span>,
                                    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>,
                                    <span class="dt">path     =</span> <span class="kw">file.path</span>(<span class="st">"_ncbi_downloads"</span>,<span class="st">"proteomes"</span>) )

Hsapiens_proteome &lt;-<span class="st"> </span>biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/read_proteome">read_proteome</a></span>(file_path, <span class="dt">format =</span> <span class="st">"fasta"</span>)

<span class="co"># remove splice variants from id</span>
gene_set &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">sapply</span>(<span class="kw">strsplit</span>(Hsapiens_proteome<span class="op">@</span>ranges<span class="op">@</span>NAMES[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>], <span class="st">"."</span>,<span class="dt">fixed =</span> <span class="ot">TRUE</span>), <span class="cf">function</span>(x) x[<span class="dv">1</span>]))

result_BM &lt;-<span class="st"> </span>biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/biomart">biomart</a></span>( <span class="dt">genes      =</span> gene_set, <span class="co"># genes were retrieved using biomartr::getGenome()</span>
                                <span class="dt">mart       =</span> <span class="st">"ENSEMBL_MART_ENSEMBL"</span>, <span class="co"># marts were selected with biomartr::getMarts()</span>
                                <span class="dt">dataset    =</span> <span class="st">"hsapiens_gene_ensembl"</span>, <span class="co"># datasets were selected with biomartr::getDatasets()</span>
                                <span class="dt">attributes =</span> <span class="kw">c</span>(<span class="st">"ensembl_gene_id"</span>,<span class="st">"ensembl_peptide_id"</span>), <span class="co"># attributes were selected with biomartr::getAttributes()</span>
                                <span class="dt">filters    =</span> <span class="st">"refseq_peptide"</span>) <span class="co"># specify what ID type was stored in the fasta file retrieved with biomartr::getGenome()</span>

result_BM </code></pre></div>
<p>The <code><a href="../reference/biomart.html">biomart()</a></code> function takes as arguments a set of genes (gene ids specified in the <code>filter</code> argument), the corresponding <code>mart</code> and <code>dataset</code>, as well as the <code>attributes</code> which shall be returned.</p>
</div>
<div id="gene-ontology" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-ontology" class="anchor"></a>Gene Ontology</h2>
<p>The <code>biomartr</code> package also enables a fast and intuitive retrieval of GO terms and additional information via the <code><a href="../reference/getGO.html">getGO()</a></code> function. Several databases can be selected to retrieve GO annotation information for a set of query genes. So far, the <code><a href="../reference/getGO.html">getGO()</a></code> function allows GO information retrieval from the <a href="http://ensemblgenomes.org/info/access/biomart">Ensembl Biomart</a> database.</p>
<p>In this example we will retrieve GO information for a set of <em>Homo sapiens</em> genes stored as <code>hgnc_symbol</code>.</p>
<div id="go-annotation-retrieval-via-biomart" class="section level3">
<h3 class="hasAnchor">
<a href="#go-annotation-retrieval-via-biomart" class="anchor"></a>GO Annotation Retrieval via BioMart</h3>
<p>The <code><a href="../reference/getGO.html">getGO()</a></code> function takes several arguments as input to retrieve GO information from BioMart. First, the scientific name of the <code>organism</code> of interest needs to be specified. Furthermore, a set of <code>gene ids</code> as well as their corresponding <code>filter</code> notation (<code>GUCA2A</code> gene ids have <code>filter</code> notation <code>hgnc_symbol</code>; see <code><a href="../reference/organismFilters.html">organismFilters()</a></code> for details) need to be specified. The <code>database</code> argument then defines the database from which GO information shall be retrieved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all elements of the data.frame</span>
<span class="kw">options</span>(<span class="dt">tibble.print_max =</span> <span class="ot">Inf</span>)
<span class="co"># search for GO terms of an example Homo sapiens gene</span>
GO_tbl &lt;-<span class="st"> </span>biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getGO">getGO</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>, 
                          <span class="dt">genes    =</span> <span class="st">"GUCA2A"</span>,
                          <span class="dt">filters  =</span> <span class="st">"hgnc_symbol"</span>)

GO_tbl</code></pre></div>
<p>Hence, for each <em>gene id</em> the resulting table stores all annotated GO terms found in <a href="http://ensemblgenomes.org/info/access/biomart">Ensembl Biomart</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#functional-annotation-retrieval-from-ensembl-biomart">Functional Annotation Retrieval from <code>Ensembl Biomart</code></a></li>
      <li><a href="#extending-biomart-using-the-new-query-system-of-the-biomartr-package">Extending <code>biomaRt</code> using the new query system of the <code>biomartr</code> package</a></li>
      <li><a href="#retrieve-marts-datasets-attributes-and-filters-with-biomartr">Retrieve marts, datasets, attributes, and filters with biomartr</a></li>
      <li><a href="#organism-specific-retrieval-of-information">Organism Specific Retrieval of Information</a></li>
      <li><a href="#construct-biomart-queries-with-biomartr">Construct BioMart queries with <code>biomartr</code></a></li>
      <li><a href="#gene-ontology">Gene Ontology</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
