<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • biomartr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biomartr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.9.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/BioMart_Examples.html">Ensembl BioMart Examples</a>
    </li>
    <li>
      <a href="../articles/Database_Retrieval.html">NCBI Database Retrieval</a>
    </li>
    <li>
      <a href="../articles/Functional_Annotation.html">Functional Annotation with biomartr</a>
    </li>
    <li>
      <a href="../articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a>
    </li>
    <li>
      <a href="../articles/Sequence_Retrieval.html">Sequence Retrieval</a>
    </li>
    <li>
      <a href="../articles/repsonse.html">UNKNOWN TITLE</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ropensci/biomartr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ropensci/biomartr/blob/master/vignettes/repsonse.Rmd"><code>vignettes/repsonse.Rmd</code></a></small>
      <div class="hidden name"><code>repsonse.Rmd</code></div>

    </div>

    
    
<div id="response-to-package-reviews" class="section level2">
<h2 class="hasAnchor">
<a href="#response-to-package-reviews" class="anchor"></a>Response to Package Reviews</h2>
<p>Dear <span class="citation">@sckott</span>,</p>
<p>I am very grateful for receiving two valuable reviews from <span class="citation">@grimbough</span> and <span class="citation">@naupaka</span>. In the following response to the reviewers, I tried to address all comments and issues and incorporated all changes into the new version of the <code>biomartr</code> package (see <a href="https://github.com/HajkD/biomartr/blob/master/NEWS.html">NEWS</a>).</p>
<p>I hope that I could address all points sufficiently so that you can consider <code>biomartr</code> to become a part of the rOpenSci package collection.</p>
<p>I would like to thank you and both reviewers for your detailed and constructive comments. It was a pleasure to work with you and I am looking forward to collaborating with the rOpenSci community in the future.</p>
<p>Kind regards, Hajk</p>
<div id="response-to-reviewer-1-grimbough" class="section level3">
<h3 class="hasAnchor">
<a href="#response-to-reviewer-1-grimbough" class="anchor"></a>Response to reviewer 1 (<span class="citation">@grimbough</span>)</h3>
<div id="general-comments" class="section level4">
<h4 class="hasAnchor">
<a href="#general-comments" class="anchor"></a>General comments</h4>
<blockquote>
<p>I actually find the package name a little misleading, since this does way more than just provide an R interface to the BioMart API. Perhaps that was the original intention for the package, but the present functionality exceeds this. It’s probably a bit late to change this though given there are existing publications referring to it.</p>
</blockquote>
<p><strong>Response</strong>: I agree. The package naming is constrained due to the history of how <code>biomartr</code> was developed in the first place. Initially, I focused on the BioMart database, but then it closed it’s service and the NCBI and ENSEMBL databases took over parts of it’s role and diversified their services as well. Another point was the review process of the <code>biomartr</code> package in the journal <code>Bioinformatics</code>. There, it was requested during the review process that the functionality of the <code>biomartr</code> package should be extended to retrieve genomes, proteomes, etc. from NCBI and ENSEMBL databases. I apologize that all this caused a misleading naming of the package, but I hope that the detailed documentation will make users aware of this shortcoming.</p>
<blockquote>
<p>One other small semantic point, which you see a lot, is that this package provides access to ‘Ensembl Biomart’, rather than accessing Biomart services in general. There are in fact many other databases that use Biomart as a way to query their data e.g. SalmonDB or Pancreas Expression Database, which this package currently doesn’t provide access to. This is particularly true since the centralised access via Biomart.org no longer exists, and it might be worth clarifying this in the README and Introductory vignette.</p>
</blockquote>
<p><strong>Response</strong>: Many thanks for pointing this out. I now made this point clear in the introductory vignette.</p>
</div>
<div id="buildinstall" class="section level4">
<h4 class="hasAnchor">
<a href="#buildinstall" class="anchor"></a>Build/Install</h4>
<blockquote>
<p>The author might be interested to know that you can use biocLite() to install from CRAN and github as well as Bioconductor, and it will sort of the dependencies from the various repos. So the installation of the CRAN version can be simplified.</p>
</blockquote>
<p><strong>Response</strong>: This is a great suggestion, and I now use the <code>biocLite()</code> way in the README.</p>
<blockquote>
<p>One other thing to note is that the importing of functions from biomaRt and Biostrings is never defined explicitly. They are mentioned in the installation instructions and their functions are accessed using the double colon operator e.g. Biostrings::readBStringSet(). However they are not declared in the NAMESPACE. Including the relevant functions as imports would be desireable. Otherwise, checking for their existence with useNamespace() and printing a helpful message if they aren’t present may also work.</p>
</blockquote>
<p><strong>Response</strong>: I fixed this now and explicitly imported <code>biomaRt</code> and <code>Biostrings</code> in the <code>NAMESPACE</code>.</p>
<blockquote>
<p>Most functions have examples, and those that I tested manually worked. However, they are all set to not run with DONTRUN flags, so it is difficult to run them in an automated fashion. This also means they aren’t run when the package is built and check, so there is a chance that some no longer execute successfully. I would certainly encourage the package maintainers to make it that more of the examples are run, although I appreciate that they probably take a long time to execute given almost all are retrieving online data sets. Running the set of tests that accompany the package to over 30 minutes for me.</p>
</blockquote>
<p><strong>Response</strong>: Thank you so much for pointing this out to me, but to be honest, I am not sure how to run these download commands while following the CRAN policy that states that examples should terminate within 5 sec. In my initial CRAN submission, I didn’t have <code>DONTRUNS</code> wrapped around examples and was asked by the CRAN maintainers to do so to comply with the 5 sec example rule. Due to this fact, I started using <code>Travis</code> to automatically check the functionality of all functions whenever I update the package. However, I am very happy for any suggestion on how to run download examples and still comply with the 5 sec CRAN policy.</p>
<blockquote>
<p>When running devtools::test() I get three failures, however in each case when I manually ran the code I got the expected output rather than an error. I haven’t had a chance to explore this further, but I have include the failure messages below for completeness.</p>
</blockquote>
<p><strong>Response</strong>: I also explored this peculiarity and fixed the unit tests.</p>
</div>
<div id="documentation-and-vignettes" class="section level4">
<h4 class="hasAnchor">
<a href="#documentation-and-vignettes" class="anchor"></a>Documentation and vignettes</h4>
<blockquote>
<p>The vignettes are generally well written and cover a variety of use cases. The vignette naming is also based around these use cases, which I guess makes sense if you approach the package with a particular research domain in mind, but is not so helpful if you a looking for how to access a specific service. For example, if I know I want to query Ensembl BioMart (which doesn’t seem unreasonable given the package name) it doesn’t feel intuitive to me that I have to look in ‘Evolutionary Transcriptomics’ to find an example of accessing that particular resource. This would be somewhat mitigated by the suggestion to include services in the function names, so it is at least easier to find the appropriate manual page.</p>
</blockquote>
<p><strong>Response</strong>: I agree and now renamed the vignette to <code>BioMart Examples</code> and adapted the content of this vignette.</p>
<blockquote>
<p>I would highly recommend making more of the vignette examples evaluated when the Rmarkdown documents are rendered. Over time the contents of the databases can change, and having static output in the vignettes can lead to confusion among users, when what they see when running a code block no longer matches what they see in the documentation. For example running the command listDatabases(db = “human”) produces a list of files that is different from what is contained in the Database_Retrieval vignette. You can also end up in the situation where examples in a vignette no longer work at all, although having a decent set of tests in the package should reduce the chances of this happening. This is something where a repository like Bioconductor, which builds all of its packages regularly on multiple platforms, can be very helpful, since you are notified if the vignette no longer executes.</p>
</blockquote>
<p><strong>Response</strong>: I agree, and now vignette examples are evaluated when the Rmarkdown documents are rendered.</p>
</div>
<div id="functionvariable-naming-general-syntax" class="section level4">
<h4 class="hasAnchor">
<a href="#functionvariable-naming-general-syntax" class="anchor"></a>Function/variable naming &amp; general syntax</h4>
<blockquote>
<p>In general the code is pretty easy to read, with sensible variable names and neatly structured code. However, the function naming within the package seems to be inconsistent. There are examples of camelCaps, snake_case and period.separated. Personally I would avoid using period.separated since there is the potential to conflict with the S3 method dispatch system, but you can find at least one style guide to support each. However, it would be nice to be consistent within the package.</p>
</blockquote>
<p><strong>Response</strong>: Thanks for pointing this out to me. In the next versions of <code>biomartr</code>, I will consistently use the <code>camelCaps</code> style for new functions.</p>
<blockquote>
<p>I also think from a user view point that some function names could perhaps be more explicit about the resource they are accessing. e.g. listDatabases() only accesses the NCBI, but all of the services this package can connect to are ultimately databases. Similarly getDatasets() is BioMart specific. This is explained in the manual pages, but it might make it more intuitive to users to have the service in the function name so you can more easily pick out the functions to use together. I feel this is particularly true when you consider functions like getGenome() actually do have the functionality to access multiple resources. This is addressed a bit by grouping related functions together in the README.</p>
</blockquote>
<p><strong>Response</strong>: I renamed the <code><a href="../reference/listDatabases.html">listDatabases()</a></code> function to <code><a href="../reference/listDatabases.html">listNCBIDatabases()</a></code> and will depreciate <code><a href="../reference/listDatabases.html">listDatabases()</a></code> in the next version of <code>biomartr</code>. Due to the documentation of the Ensembl Biomart query notation <code><a href="../reference/getDatasets.html">biomartr::getDatasets()</a></code> is now more clearly defined.</p>
</div>
<div id="console-messages" class="section level4">
<h4 class="hasAnchor">
<a href="#console-messages" class="anchor"></a>Console messages</h4>
<blockquote>
<p>There are quite a few instances of cat() and print() scattered throughout the code to write messages to the user e.g. in getGenome.R, getGFF.R etc. I would recommend changing these to message() so they can be suppressed if desired. The cat() usage is often in conjunction with sink() and setwd(). I am more in favour of writeLines() over sink()&amp;cat() but I that’s personal preference rather than anything concrete. However, I would avoid using setwd() if possible, since this changes things in the global environment and if your function fails for any reason the working directory has now changed for the user. I would either pass the path to the file writing function, or if you really want to change directory, use on.exit(setwd(“my/original/dir”)) to ensure it is changed back however the current function exits.</p>
</blockquote>
<p><strong>Response</strong>: I agree and now consistently use <code>message()</code>. I also use the <code>file.path()</code> instead of changing the working directory.</p>
</div>
<div id="is-there-code-duplication-in-the-package-that-should-be-reduced" class="section level4">
<h4 class="hasAnchor">
<a href="#is-there-code-duplication-in-the-package-that-should-be-reduced" class="anchor"></a>Is there code duplication in the package that should be reduced?</h4>
<blockquote>
<p>There are a few instances where variables such as file names are created multiple times throughout the same function. For example file.path(tempdir(), “_ncbi_downloads“,”listDatabases.txt“) appears six times in listDatabases.R. It’s only a very minor detail, but having it defined only once may save hassle in the future if you ever decide to change that particular file name.</p>
</blockquote>
<p><strong>Response</strong>: I now store <code>file.path(tempdir(), "_ncbi_downloads", "listDatabases.txt")</code> in the variable <code>db.file.name</code>.</p>
<blockquote>
<p>getGenome() has a degree of code replication between the paths followed when db = “ensemblgenomes” and db = “ensembl”. I haven’t thought about it too much, but I wonder if this function could be modularised with two sub-functions, one of which handles refseq and genbank queries, and one which handles the two ensembl queries? Similarly there is a lot of overlap between getENSEMBL.Seq.R and getENSEMBLGENOMES.Seq.R. I wonder if these could be combined, and the requested host used to select the few lines that are different?</p>
</blockquote>
<p><strong>Response</strong>: Due to the differences in the internal folder structure of the ENSEMBL and ENSEMBLGENOMES servers I had to implement special cases and thus this code redundancy was on purpose to have a better overview for code maintainence.</p>
</div>
</div>
<div id="response-to-reviewer-2-naupaka" class="section level3">
<h3 class="hasAnchor">
<a href="#response-to-reviewer-2-naupaka" class="anchor"></a>Response to reviewer 2 (<span class="citation">@naupaka</span>)</h3>
<div id="review-comments" class="section level4">
<h4 class="hasAnchor">
<a href="#review-comments" class="anchor"></a>Review Comments</h4>
<blockquote>
<p>I think it would be nice to clarify in the README and elsewhere that the queries work, for the most part (as far as I understand it), against NCBI or EMSEMBL, and not from any of the many other sources that have a BioMart interface (as <span class="citation">@grimbough</span> pointed out). For example, I am interested in getting data from the BioMart interface to JGI’s Phytozome, since they have a 3.0 version of the Populus trichocarpa genome (the NCBI version is still at 2.0), but that is not possible, I believe, with this package (feature request ).</p>
</blockquote>
<p><strong>Response</strong>: I agree and I changed it according to <span class="citation">@grimbough</span>’s point. The JGI’s Phytozome request is taken (<a href="https://github.com/HajkD/biomartr/issues/9" class="uri">https://github.com/HajkD/biomartr/issues/9</a>) and since I am working with plant genomes myself I will try my best to provide some interface functions :D .</p>
<blockquote>
<p>Another thing that I didn’t see, but I think would be useful, would be to modify listGenomes() to return a data frame where the first column is is the species’ Latin binomial and the second column is which database each species’ genome is associated with. Otherwise, it’s tricky to figure out which to set for the db = parameter when using getGenome().</p>
</blockquote>
<p><strong>Response</strong>: This is a functionality extension request. I am happy to implement it for the next <code>biomartr</code> version and opened up an <a href="https://github.com/HajkD/biomartr/issues/10">extension request</a>.</p>
</div>
<div id="readme" class="section level4">
<h4 class="hasAnchor">
<a href="#readme" class="anchor"></a>README</h4>
<blockquote>
<p>No Travis badge in README.</p>
</blockquote>
<p><strong>Response</strong>: Travis badge is now included in README.</p>
<blockquote>
<p>While all exported functions have roxygen-generated documentation, some internal functions have no comments or documentation, e.g. connected.to.internet.R, exists.ftp.file.R, or get.emsembl.info.R. You could add some short roxygen comments and add #’ <span class="citation">@noRd</span> as recommended in the ROpenSci package guidelines, or just have some non-roxygen style comments in there in case others would like to better understand what’s going on.</p>
</blockquote>
<p><strong>Response</strong>: Many thanks for pointing this out to me. I now commented all internal functions and used <code>#' @noRd</code> to mark that it is an internal function.</p>
<blockquote>
<p>I also agree with <span class="citation">@grimbough</span> that it would be ideal if at least some of the example code in the documentation could actually be run. The output from devtools::run_examples() suggests that there isn’t any code in any of the examples for any of the functions that is actually run. Obviously, it wouldn’t make sense to run download.database.all(name = “nr”, path = “nr”) because that could take days, but surely there are some queries that would be fast enough to run in a reasonable timeframe?</p>
</blockquote>
<p><strong>Response</strong>: I agree, but please see my response to <span class="citation">@grimbough</span> concerning the 5s CRAN policy.</p>
</div>
<div id="code" class="section level4">
<h4 class="hasAnchor">
<a href="#code" class="anchor"></a>Code</h4>
<blockquote>
<p>I was a bit confused by which functions did what, in terms of which database(s) they query, and where exactly datasets were coming from. For each download, this information is all very helpfully documented in the doc_*.txt files in the _ncbi_downloads directory, but I felt like it was a bit of a hunt to figure out what had happened. Maybe this is just because I am still not 100% clear on which databases this package interfaces with. This is related to my earlier comment about being confused about how to figure out which db to get a genome from if you don’t know ahead of time.</p>
</blockquote>
<p><strong>Response</strong>: As stated above, I now clearly write the databases <code>biomartr</code> is interfacing in the vignettes and also changed the function name from <code><a href="../reference/listDatabases.html">listDatabases()</a></code> ti <code><a href="../reference/listDatabases.html">listNCBIDatabases()</a></code>. I hope that it is less confusing now.</p>
<blockquote>
<p>ROpenSci’s packaging guide suggests the use of message() and/or warning() instead of print() (e.g. lines 144 and 291 of getProteome.R)</p>
</blockquote>
<p><strong>Response</strong>: I now consistently use <code>message() and/or warning()</code> instead of <code>print()</code> or <code>cat()</code>.</p>
</div>
<div id="tests" class="section level4">
<h4 class="hasAnchor">
<a href="#tests" class="anchor"></a>Tests</h4>
<blockquote>
<p>A number of the tests fail for me when run with either R CMD CHECK or devtools::check(). Not sure if it is appropriate to put all that output here, but I figured maybe it would be helpful for diagnosing the problem. I think these errors are related to the occasional ‘weird server error’ messages I was getting.</p>
</blockquote>
<p><strong>Response</strong>: I fixed the corrupt unit tests and now <code><a href="http://www.rdocumentation.org/packages/devtools/topics/check">devtools::check(document = FALSE)</a></code> does no throw errors anymore.</p>
<blockquote>
<p>Looks like there is a non-ASCII character in the copied tibble output in one of the verbatim blocks in the Functional Annotation vignette (non-executing).</p>
</blockquote>
<p><strong>Response</strong>: Thank you so much for detecting this non-ASCII character. It is fixed it now.</p>
<blockquote>
<p>Need line break before line 33 in Introduction.Rmd, or markdown header doesn’t render properly.</p>
</blockquote>
<p><strong>Response</strong>: I included the line break.</p>
<blockquote>
<p>Repeated “No encoding supplied” warnings</p>
</blockquote>
<p><strong>Response</strong>: The encoding is fixed now.</p>
<blockquote>
<p>Sometimes when I would try to run a command, I’d get errors like this. I was connected to the Internet at the time, and similar commands run before and afterwards worked fine. Sometimes the same command, run a second time, worked fine even though it failed at first. Maybe something about the timeout before waiting for a server response could be tweaked to reduce the likelihood of premature failure?</p>
</blockquote>
<p><strong>Response</strong>: Thank you so much for pointing this server timeout issue out to me. I now implemented a customized download function which uses the best download tool for the underlying operating system. I hope that this issue is now resolved.</p>
<p>===============</p>
<p>Dear <span class="citation">@sckott</span>,</p>
<p>I addressed all comments and issues and incorporated all changes into the new version of the <code>biomartr</code> package.</p>
<p>I hope that I could address all points sufficiently so that you can consider <code>biomartr</code> to become a part of the <code>rOpenSci</code> package collection.</p>
<p>Please find my detailed response below.</p>
<p>Thank you so much for all your support.</p>
<p>Kind regards,</p>
<p>Hajk</p>
<p>P.S. I just want to make you aware that the Ensembl Services are <a href="https://ensembl.github.io/ensembl-status/index.html">currently not available</a>. Thus, until the Ensembl servers is up again some of the <code>biomartr</code> functions will not run properly.</p>
<blockquote>
<p>For running egs, I suggest MAYBE running egs wrapped in dontrun on Travis by adding –run-dontrun to your check args (since taking a longish time on travis I assume is not a big deal) - I’d have one or two examples for each function not wrapped in dontrun when you know they will run quickly. Note: I said MAYBE since check already takes a very long time, but still would be nice to make sure egs always work</p>
</blockquote>
<p><strong>Response</strong>: I tried to solve the running time issue by choosing example test cases that use very small files. It should be better now. Please also see my response to the test coverage and <code>R CMD check</code>.</p>
<blockquote>
<p>output of goodpractice::gp() was pointed out and there’s some things in there which I think you should address: sapply usage: replace with vapply/lapply it’s a bit of work, but I highly suggest tidying code and documentation to 80 line width so you don’t get text running off the page on the CRAN manual (e.g., <a href="https://cran.rstudio.com/web/packages/biomartr/biomartr.pdf" class="uri">https://cran.rstudio.com/web/packages/biomartr/biomartr.pdf</a>), and it’s easier to see in a narrow width editor setup replace usage of 1:length(…), 1:nrow(…), 1:ncol(…), 1:NROW(…) and 1:NCOL(…) with eg., seq_along/seq_len</p>
</blockquote>
<p><strong>Response</strong>: I agree and now replaced all <code>sapply()</code> commands by <code>lapply()</code>, reformatted all code and documentation to 80 line width, and replace usage of <code>1:length(...)</code>, <code>1:nrow(...)</code>, <code>1:ncol(...)</code>, <code>1:NROW(...)</code> and <code>1:NCOL(...)</code> with <code>seq_len()</code>.</p>
<blockquote>
<p>Add a package level manual file - so that users can do ?biomartr to get to the man pages easily</p>
</blockquote>
<p><strong>Response</strong>: A package level manual file is now included and as suggested can be accessed via <code><a href="../reference/biomartr-package.html">?biomartr</a></code>.</p>
<blockquote>
<p>looks like test coverage is about 47% - would be good to get that % up - not necessary now, but in the future.</p>
</blockquote>
<p><strong>Response</strong>: I now spent a significant amount of time on increasing the test coverage and will continue to improve it in the future.</p>
<blockquote>
<p>R cmd check takes a long time, just for ease of quick turn around time, it’d be wise to make that time much less - I assume most of the time is in tests, but haven’t looked into how to speed that up</p>
</blockquote>
<p><strong>Response</strong>: Most tests cover the functionality of data retrieval functions, so to reduce <code>R cmd check</code> time I now choose small file downloads for unit testing.</p>
<blockquote>
<p>also related to tests, but not about speed: looks like many of your test files aren’t doing much testing using testthat - that is, you run examples, but then don’t test their output (e.g., <a href="https://github.com/HajkD/biomartr/blob/master/tests/testthat/test-is.genome.available.R#L12-L36" class="uri">https://github.com/HajkD/biomartr/blob/master/tests/testthat/test-is.genome.available.R#L12-L36</a>) I imagine the majority of time in tests right now is doing data requests - adding test expectations shouldn’t add a lot more time</p>
</blockquote>
<p><strong>Response</strong>: I absolutely agree and see where your confusion comes from. I now commented and specified the unit tests more clearly, so that no confusion exists on what part of the function is actually tested.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#response-to-package-reviews">Response to Package Reviews</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
