<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Genome Assembly Stats Retrieval — getAssemblyStats • biomartr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Genome Assembly Stats Retrieval — getAssemblyStats" />

<meta property="og:description" content="Main genome assembly stats retrieval function for an organism
of interest. By specifying the scientific name of an organism of interest the
corresponding  genome assembly stats file storing the assembly statistics of 
the organism of interest can be downloaded and stored locally. 
Genome assembly stats files can be retrieved from several databases." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biomartr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.8.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/BioMart_Examples.html">Ensembl BioMart Examples</a>
    </li>
    <li>
      <a href="../articles/Database_Retrieval.html">NCBI Database Retrieval</a>
    </li>
    <li>
      <a href="../articles/Functional_Annotation.html">Functional Annotation with biomartr</a>
    </li>
    <li>
      <a href="../articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a>
    </li>
    <li>
      <a href="../articles/Sequence_Retrieval.html">Sequence Retrieval</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ropensci/biomartr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genome Assembly Stats Retrieval</h1>
    <small>Source: <a href='https://github.com/ropensci/biomartr/blob/master/R/getAssemblyStats.R'><code>R/getAssemblyStats.R</code></a></small>
    <div class="hidden name"><code>getAssemblyStats.Rd</code></div>
    </div>

    
    <p>Main genome assembly stats retrieval function for an organism
of interest. By specifying the scientific name of an organism of interest the
corresponding  genome assembly stats file storing the assembly statistics of 
the organism of interest can be downloaded and stored locally. 
Genome assembly stats files can be retrieved from several databases.</p>
    

    <pre class="usage"><span class='fu'>getAssemblyStats</span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>, <span class='no'>organism</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"download"</span>, <span class='kw'>path</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='st'>"_ncbi_downloads"</span>,
  <span class='st'>"genomeassembly_stats"</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>db</th>
      <td><p>a character string specifying the database from which the genome 
shall be retrieved:</p><ul>
<li><p><code>db = "refseq"</code></p></li>
<li><p><code>db = "genbank"</code></p></li>
<li><p><code>db = "ensembl"</code></p></li>
<li><p><code>db = "ensemblgenomes"</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p>a character string specifying the scientific name of the 
organism of interest, e.g. <code>organism = "Homo sapiens"</code>.</p></td>
    </tr>
    <tr>
      <th>reference</th>
      <td><p>a logical value indicating whether or not a genome shall be downloaded if it isn't marked in the database as either a reference genome or a representative genome.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>shall only the file be retrieved (default) 
<code>type = "download"</code> or should the corresponding file be downloaded and 
subsequently be imported <code>type = "import"</code>.</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a character string specifying the location (a folder) in
which the corresponding file shall be stored. Default is 
<code>path</code> = <code>file.path("_ncbi_downloads","genomeassembly_stats")</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>File path to downloaded genome assembly stats file.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Internally this function loads the the overview.txt file from NCBI:</p>
<p>refseq: ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/</p>
<p>genbank: ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/</p> 
<p>to retrieve available scientific names of organisms and creates a directory
'_ncbi_downloads/genomeassembly_stats' to store
the Genome Assembly Stats of interest as text file for future processing.
In case the corresponding fasta file already exists within the
'_ncbi_downloads/genomeassembly_stats' folder and is
accessible within the workspace, no download process will be performed.</p>
<p>An example genome assembly stats file can be found here:
ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/
GCF_000001405.36_GRCh38.p10/GCF_000001405.36_GRCh38.p10_assembly_stats.txt.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='getGenome.html'>getGenome</a></code>, <code><a href='getProteome.html'>getProteome</a></code>, <code><a href='getCDS.html'>getCDS</a></code>,
<code><a href='getGFF.html'>getGFF</a></code>, <code><a href='getRNA.html'>getRNA</a></code>, <code><a href='meta.retrieval.html'>meta.retrieval</a></code>, 
<code><a href='read_assemblystats.html'>read_assemblystats</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># download the genome assembly stats file of Saccharomyces cerevisiae</span>
<span class='co'># from NCBI RefSeq</span>
<span class='co'># and store the corresponding genome file in </span>
<span class='co'># '_ncbi_downloads/genomeassembly_stats'</span>
<span class='no'>file_path</span> <span class='kw'>&lt;-</span> <span class='fu'>getAssemblyStats</span>( <span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>,
                 <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Saccharomyces cerevisiae"</span>,
                 <span class='kw'>path</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='st'>"_ncbi_downloads"</span>,<span class='st'>"genomeassembly_stats"</span>))
<span class='co'># import the raw file as it is downloaded</span>
<span class='no'>Scerevisiae.stats</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='read_assemblystats.html'>read_assemblystats</a></span>(<span class='no'>file_path</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"raw"</span>)

<span class='co'># download the genome assembly stats file of Saccharomyces cerevisiae</span>
<span class='co'># from NCBI RefSeq </span>
<span class='co'># and import overall statistics of the genome assembly</span>
<span class='no'>Scerevisiae.stats.import</span> <span class='kw'>&lt;-</span> <span class='fu'>getAssemblyStats</span>( <span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>,
                 <span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Saccharomyces cerevisiae"</span>,
                 <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"import"</span>,
                 <span class='kw'>path</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='st'>"_ncbi_downloads"</span>,<span class='st'>"genomeassembly_stats"</span>))
<span class='co'># }</span><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Hajk-Georg Drost

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

