<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Genome Annotation Retrieval (GTF) — getGTF • biomartr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Genome Annotation Retrieval (GTF) — getGTF" />
<meta property="og:description" content="Main retrieval function for GTF files of an
organism of interest. By specifying the scientific name of an organism of
interest the corresponding GTF file storing the annotation  for the organism
of interest can be downloaded and stored locally. GTF files can be retrieved
from several databases." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biomartr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/BioMart_Examples.html">Ensembl BioMart Examples</a>
    </li>
    <li>
      <a href="../articles/Database_Retrieval.html">NCBI Database Retrieval</a>
    </li>
    <li>
      <a href="../articles/Functional_Annotation.html">Functional Annotation with biomartr</a>
    </li>
    <li>
      <a href="../articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a>
    </li>
    <li>
      <a href="../articles/Sequence_Retrieval.html">Sequence Retrieval</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ropensci/biomartr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genome Annotation Retrieval (GTF)</h1>
    <small class="dont-index">Source: <a href='https://github.com/ropensci/biomartr/blob/master/R/getGTF.R'><code>R/getGTF.R</code></a></small>
    <div class="hidden name"><code>getGTF.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Main retrieval function for GTF files of an
organism of interest. By specifying the scientific name of an organism of
interest the corresponding GTF file storing the annotation  for the organism
of interest can be downloaded and stored locally. GTF files can be retrieved
from several databases.</p>
    </div>

    <pre class="usage"><span class='fu'>getGTF</span><span class='op'>(</span>
  db <span class='op'>=</span> <span class='st'>"ensembl"</span>,
  <span class='va'>organism</span>,
  remove_annotation_outliers <span class='op'>=</span> <span class='cn'>FALSE</span>,
  path <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"ensembl"</span>, <span class='st'>"annotation"</span><span class='op'>)</span>,
  assembly_type <span class='op'>=</span> <span class='st'>"toplevel"</span>,
  release <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>db</th>
      <td><p>a character string specifying the database from which the genome
shall be retrieved:</p><ul>
<li><p><code>db = "ensembl"</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p>a character string specifying the scientific name of the
organism of interest, e.g. <code>organism = "Homo sapiens"</code>.</p></td>
    </tr>
    <tr>
      <th>remove_annotation_outliers</th>
      <td><p>shall outlier lines be removed from the input <code>annotation_file</code>?
If yes, then the initial <code>annotation_file</code> will be overwritten and the removed outlier lines will be stored at <code><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></code>
for further exploration.</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a character string specifying the location (a folder) in which
the corresponding annotation file shall be stored. Default is
<code>path = file.path("ensembl","annotation")</code>.</p></td>
    </tr>
    <tr>
      <th>assembly_type</th>
      <td><p>a character string specifying from which assembly type the genome
shall be retrieved from (ensembl only, else this argument is ignored):
Default is
<code>assembly_type = "toplevel")</code>.
This will give you all multi-chromosomes (copies of the same chromosome with small variations).
As an example the toplevel fasta genome in human is over 70 GB uncompressed.
To get primary assembly with 1 chromosome variant per chromosome:
<code>assembly_type = "primary_assembly")</code>.
As an example, the  primary_assembly fasta genome in human is only a few GB uncompressed:</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p>a numeric, the database release version of ENSEMBL (<code>db = "ensembl"</code>). Default is <code>release = NULL</code> meaning
that the most recent database version is used. <code>release = 75</code> would for human would give the stable
GRCh37 release in ensembl. Value must be &gt; 46, since ensembl did not structure their data
if the standard format before that.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>File path to downloaded annotation file.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Internally this function loads the the overview.txt file from ENSEMBL:
and creates a directory 'ensembl/annotation' to store
the genome of interest as fasta file for future processing.
In case the corresponding fasta file already exists within the
'ensembl/annotation' folder and is accessible within the workspace,
no download process will be performed.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='getProteome.html'>getProteome</a></code>, <code><a href='getCDS.html'>getCDS</a></code>,
<code><a href='getGenome.html'>getGenome</a></code>, <code><a href='getRNA.html'>getRNA</a></code>, <code><a href='getRepeatMasker.html'>getRepeatMasker</a></code>,
<code><a href='getAssemblyStats.html'>getAssemblyStats</a></code>, <code><a href='meta.retrieval.html'>meta.retrieval</a></code>,
<code><a href='getGFF.html'>getGFF</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Hajk-Georg Drost</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># download the annotation of Homo sapiens from ensembl</span>
<span class='co'># and store the corresponding genome file in 'ensembl/annotation'</span>
<span class='fu'>getGTF</span><span class='op'>(</span>db            <span class='op'>=</span> <span class='st'>"ensembl"</span>,
       organism      <span class='op'>=</span> <span class='st'>"Homo sapiens"</span>,
       path          <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"ensembl"</span>,<span class='st'>"annotation"</span><span class='op'>)</span><span class='op'>)</span>

<span class='fu'>getGTF</span><span class='op'>(</span>db            <span class='op'>=</span> <span class='st'>"ensembl"</span>,
       organism      <span class='op'>=</span> <span class='st'>"Homo sapiens"</span>,
       path          <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='st'>"ensembl"</span>,<span class='st'>"annotation"</span><span class='op'>)</span>,
       assembly_type <span class='op'>=</span> <span class='st'>"primary_assembly"</span><span class='op'>)</span>

<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


