<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ biomartr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biomartr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/BioMart_Examples.html">Ensembl BioMart Examples</a>
    </li>
    <li>
      <a href="../articles/Database_Retrieval.html">NCBI Database Retrieval</a>
    </li>
    <li>
      <a href="../articles/Functional_Annotation.html">Functional Annotation with biomartr</a>
    </li>
    <li>
      <a href="../articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a>
    </li>
    <li>
      <a href="../articles/Sequence_Retrieval.html">Sequence Retrieval</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ropensci/biomartr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="biomart.html">biomart()</a></code> </p>
        </td>
        <td><p>Main BioMart Query Function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="biomartr-package.html">biomartr-package</a></code> </p>
        </td>
        <td><p>Genomic Data Retrieval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_annotation_biomartr.html">check_annotation_biomartr()</a></code> </p>
        </td>
        <td><p>Check whether an annotation file contains outlier lines</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clean.retrieval.html">clean.retrieval()</a></code> </p>
        </td>
        <td><p>Format <code>meta.retrieval</code> output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="download.database.html">download.database()</a></code> </p>
        </td>
        <td><p>Download a NCBI Database to Your Local Hard Drive</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="download.database.all.html">download.database.all()</a></code> </p>
        </td>
        <td><p>Download all elements of an NCBI databse</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAssemblyStats.html">getAssemblyStats()</a></code> </p>
        </td>
        <td><p>Genome Assembly Stats Retrieval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAttributes.html">getAttributes()</a></code> </p>
        </td>
        <td><p>Retrieve All Available Attributes for a Specific Dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getCDS.html">getCDS()</a></code> </p>
        </td>
        <td><p>Coding Sequence Retrieval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getCDSSet.html">getCDSSet()</a></code> </p>
        </td>
        <td><p>CDS retrieval of multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getCollection.html">getCollection()</a></code> </p>
        </td>
        <td><p>Retrieve a Collection: Genome, Proteome, CDS, RNA, GFF, Repeat Masker, AssemblyStats</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getCollectionSet.html">getCollectionSet()</a></code> </p>
        </td>
        <td><p>Retrieve a Collection: Genome, Proteome, CDS, RNA, GFF, Repeat Masker, AssemblyStats of multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDatasets.html">getDatasets()</a></code> </p>
        </td>
        <td><p>Retrieve All Available Datasets for a BioMart Database</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getENSEMBL.Seq.html">getENSEMBL.Seq()</a></code> </p>
        </td>
        <td><p>Helper function for retrieving biological sequence files from ENSEMBL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getENSEMBL.gtf.html">getENSEMBL.gtf()</a></code> </p>
        </td>
        <td><p>Helper function for retrieving gff files from ENSEMBL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getENSEMBLGENOMESInfo.html">getENSEMBLGENOMESInfo()</a></code> </p>
        </td>
        <td><p>Retrieve ENSEMBLGENOMES info file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getENSEMBLInfo.html">getENSEMBLInfo()</a></code> </p>
        </td>
        <td><p>Retrieve ENSEMBL info file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getFilters.html">getFilters()</a></code> </p>
        </td>
        <td><p>Retrieve All Available Filters for a Specific Dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGENOMEREPORT.html">getGENOMEREPORT()</a></code> </p>
        </td>
        <td><p>Retrieve NCBI GENOME_REPORTS file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGFF.html">getGFF()</a></code> </p>
        </td>
        <td><p>Genome Annotation Retrieval (GFF3)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGFFSet.html">getGFFSet()</a></code> </p>
        </td>
        <td><p>GFF retrieval of multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGO.html">getGO()</a></code> </p>
        </td>
        <td><p>Gene Ontology Query</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGTF.html">getGTF()</a></code> </p>
        </td>
        <td><p>Genome Annotation Retrieval (GTF)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGenome.html">getGenome()</a></code> </p>
        </td>
        <td><p>Genome Retrieval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGenomeSet.html">getGenomeSet()</a></code> </p>
        </td>
        <td><p>Genome Retrieval of multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGroups.html">getGroups()</a></code> </p>
        </td>
        <td><p>Retrieve available groups for a kingdom of life (only available for NCBI RefSeq and NCBI Genbank)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getKingdomAssemblySummary.html">getKingdomAssemblySummary()</a></code> </p>
        </td>
        <td><p>Retrieve and summarise the assembly_summary.txt files from 
NCBI for all kingdoms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getKingdoms.html">getKingdoms()</a></code> </p>
        </td>
        <td><p>Retrieve available kingdoms of life</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMarts.html">getMarts()</a></code> </p>
        </td>
        <td><p>Retrieve information about available Ensembl Biomart databases</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMetaGenomeAnnotations.html">getMetaGenomeAnnotations()</a></code> </p>
        </td>
        <td><p>Retrieve annotation *.gff files for metagenomes from NCBI Genbank</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMetaGenomeSummary.html">getMetaGenomeSummary()</a></code> </p>
        </td>
        <td><p>Retrieve the assembly_summary.txt file from NCBI genbank metagenomes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMetaGenomes.html">getMetaGenomes()</a></code> </p>
        </td>
        <td><p>Retrieve metagenomes from NCBI Genbank</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getProteome.html">getProteome()</a></code> </p>
        </td>
        <td><p>Proteome Retrieval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getProteomeSet.html">getProteomeSet()</a></code> </p>
        </td>
        <td><p>Proteome retrieval of multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRNA.html">getRNA()</a></code> </p>
        </td>
        <td><p>RNA Sequence Retrieval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRNASet.html">getRNASet()</a></code> </p>
        </td>
        <td><p>RNA Retrieval of multiple species</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getReleases.html">getReleases()</a></code> </p>
        </td>
        <td><p>Retrieve available database releases or versions 
of ENSEMBL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRepeatMasker.html">getRepeatMasker()</a></code> </p>
        </td>
        <td><p>Repeat Masker Retrieval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSummaryFile.html">getSummaryFile()</a></code> </p>
        </td>
        <td><p>Helper function to retrieve the assembly_summary.txt file from NCBI</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is.genome.available.html">is.genome.available()</a></code> </p>
        </td>
        <td><p>Check Genome Availability</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="listDatabases.html">listDatabases()</a></code> <code><a href="listDatabases.html">listNCBIDatabases()</a></code> </p>
        </td>
        <td><p>Retrieve a List of Available NCBI Databases for Download</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="listGenomes.html">listGenomes()</a></code> </p>
        </td>
        <td><p>List All Available Genomes either by kingdom, group, or subgroup</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="listGroups.html">listGroups()</a></code> </p>
        </td>
        <td><p>List number of available genomes in each taxonomic group</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="listKingdoms.html">listKingdoms()</a></code> </p>
        </td>
        <td><p>List number of available genomes in each kingdom of life</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="listMetaGenomes.html">listMetaGenomes()</a></code> </p>
        </td>
        <td><p>List available metagenomes on NCBI Genbank</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meta.retrieval.html">meta.retrieval()</a></code> </p>
        </td>
        <td><p>Perform Meta-Genome Retrieval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meta.retrieval.all.html">meta.retrieval.all()</a></code> </p>
        </td>
        <td><p>Perform Meta-Genome Retrieval of all organisms in all kingdoms of life</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organismAttributes.html">organismAttributes()</a></code> </p>
        </td>
        <td><p>Retrieve Ensembl Biomart attributes for a query organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organismBM.html">organismBM()</a></code> </p>
        </td>
        <td><p>Retrieve Ensembl Biomart marts and datasets for a query organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organismFilters.html">organismFilters()</a></code> </p>
        </td>
        <td><p>Retrieve Ensembl Biomart filters for a qyery organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_assemblystats.html">read_assemblystats()</a></code> </p>
        </td>
        <td><p>Import Genome Assembly Stats File</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_cds.html">read_cds()</a></code> </p>
        </td>
        <td><p>Import CDS as Biostrings or data.table object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_genome.html">read_genome()</a></code> </p>
        </td>
        <td><p>Import Genome Assembly as Biostrings or data.table object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_gff.html">read_gff()</a></code> </p>
        </td>
        <td><p>Import GFF File</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_proteome.html">read_proteome()</a></code> </p>
        </td>
        <td><p>Import Proteome as Biostrings or data.table object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_rm.html">read_rm()</a></code> </p>
        </td>
        <td><p>Import Repeat Masker output file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_rna.html">read_rna()</a></code> </p>
        </td>
        <td><p>Import RNA as Biostrings or data.table object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refseqOrganisms.html">refseqOrganisms()</a></code> </p>
        </td>
        <td><p>Retrieve All Organism Names Stored on refseq</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary_cds.html">summary_cds()</a></code> </p>
        </td>
        <td><p>Retrieve summary statistics for a coding sequence (CDS) file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary_genome.html">summary_genome()</a></code> </p>
        </td>
        <td><p>Retrieve summary statistics for a genome assembly file</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


