<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform Meta-Genome Retrieval — meta.retrieval • biomartr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Perform Meta-Genome Retrieval — meta.retrieval" />

<meta property="og:description" content="Download genomes, proteomes, cds, gff, rna, or assembly stats 
files of all species within a kingdom of life." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biomartr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/BioMart_Examples.html">Ensembl BioMart Examples</a>
    </li>
    <li>
      <a href="../articles/Database_Retrieval.html">NCBI Database Retrieval</a>
    </li>
    <li>
      <a href="../articles/Functional_Annotation.html">Functional Annotation with biomartr</a>
    </li>
    <li>
      <a href="../articles/MetaGenome_Retrieval.html">Meta-Genome Retrieval</a>
    </li>
    <li>
      <a href="../articles/Sequence_Retrieval.html">Sequence Retrieval</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ropensci/biomartr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform Meta-Genome Retrieval</h1>
    <small class="dont-index">Source: <a href='https://github.com/ropensci/biomartr/blob/master/R/meta.retrieval.R'><code>R/meta.retrieval.R</code></a></small>
    <div class="hidden name"><code>meta.retrieval.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Download genomes, proteomes, cds, gff, rna, or assembly stats 
files of all species within a kingdom of life.</p>
    
    </div>

    <pre class="usage"><span class='fu'>meta.retrieval</span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>, <span class='no'>kingdom</span>, <span class='kw'>group</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"genome"</span>,
  <span class='kw'>restart_at_last</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>combine</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>db</th>
      <td><p>a character string specifying the database from which the genome 
shall be retrieved:</p>
<ul>
<li><p><code>db = "refseq"</code></p></li>
<li><p><code>db = "genbank"</code></p></li>
<li><p><code>db = "emsembl"</code></p></li>
<li><p><code>db = "ensemblgenomes"</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>kingdom</th>
      <td><p>a character string specifying the kingdom of the organisms 
of interest, e.g.</p>
<ul>
<li><p>For <code>NCBI RefSeq</code>:</p><ul>
<li><p><code>kingdom = "archaea"</code></p></li>
<li><p><code>kingdom = "bacteria"</code></p></li>
<li><p><code>kingdom = "fungi"</code></p></li>
<li><p><code>kingdom = "invertebrate"</code></p></li>
<li><p><code>kingdom = "plant"</code></p></li>
<li><p><code>kingdom = "protozoa"</code></p></li>
<li><p><code>kingdom = "viral"</code></p></li>
<li><p><code>kingdom = "vertebrate_mammalian"</code></p></li>
<li><p><code>kingdom = "vertebrate_other"</code></p></li>
</ul></li>
<li><p>For <code>NCBI Genbank</code>:</p><ul>
<li><p><code>kingdom = "archaea"</code></p></li>
<li><p><code>kingdom = "bacteria"</code></p></li>
<li><p><code>kingdom = "fungi"</code></p></li>
<li><p><code>kingdom = "invertebrate"</code></p></li>
<li><p><code>kingdom = "plant"</code></p></li>
<li><p><code>kingdom = "protozoa"</code></p></li>
<li><p><code>kingdom = "vertebrate_mammalian"</code></p></li>
<li><p><code>kingdom = "vertebrate_other"</code></p></li>
</ul></li>
<li><p>For <code>ENSEMBL</code>:</p><ul>
<li><p><code>kingdom = "Ensembl"</code></p></li>
</ul></li>
<li><p>For <code>ENSEMBLGENOMES</code></p><ul>
<li><p><code>kingdom = "EnsemblBacteria"</code></p></li>
<li><p><code>kingdom = "EnsemblFungi"</code></p></li>
<li><p><code>kingdom = "EnsemblMetazoa"</code></p></li>
<li><p><code>kingdom = "EnsemblPlants"</code></p></li>
<li><p><code>kingdom = "EnsemblProtists"</code></p></li>
</ul></li>
</ul>

<p>Available kingdoms can be retrieved with <code><a href='getKingdoms.html'>getKingdoms</a></code>.</p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>only species belonging to this subgroup will be downloaded. 
Groups can be retrieved with <code><a href='getGroups.html'>getGroups</a></code>.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>type of sequences that shall be retrieved. Options are:</p>
<ul>
<li><p><code>type = "genome"</code> :
 (for genome assembly retrieval; see also <code><a href='getGenome.html'>getGenome</a></code>),</p></li>
<li><p><code>type = "proteome"</code> :
 (for proteome retrieval; see also <code><a href='getProteome.html'>getProteome</a></code>),</p></li>
<li><p><code>type = "cds"</code> :
 (for coding sequence retrieval; see also <code><a href='getCDS.html'>getCDS</a></code>),</p></li>
<li><p><code>type = "gff"</code> :
(for annotation file retrieval in gff format; see also <code><a href='getGFF.html'>getGFF</a></code>),</p></li>
<li><p><code>type = "gtf"</code> :
(for annotation file retrieval in gtf format (only for ensembl and
 ensemblgenomes); see also <code><a href='getGTF.html'>getGTF</a></code>)</p></li>
<li><p><code>type = "rna"</code> :
 (for RNA file retrieval in fasta format; see also <code><a href='getRNA.html'>getRNA</a></code>),</p></li>
<li><p><code>type = "rm"</code> :
 (for Repeat Masker output file retrieval; see also 
 <code><a href='getRepeatMasker.html'>getRepeatMasker</a></code>),</p></li>
<li><p><code>type = "assemblystats"</code> :
 (for genome assembly quality stats file retrieval; 
 see also <code><a href='getAssemblyStats.html'>getAssemblyStats</a></code>).</p></li>
</ul></td>
    </tr>
    <tr>
      <th>restart_at_last</th>
      <td><p>a logical value indicating whether or not <code>meta.retrieval</code> should pick up at the last species when re-running the function.</p><ul>
<li><p>If <code>restart_at_last = TRUE</code> (Default) then <code>meta.retrieval</code> will skip all organisms that are already present in the folder 
and will start downloading all remaining species. However, this way <code>meta.wretrieval</code> will not be able to check whether
already downloaded organism files are corrupted or not by checking the md5 checksum.</p></li>
<li><p>If <code>restart_at_last = FALSE</code> then <code>meta.retrieval</code> will start from the beginning and crawl through already downloaded
organism files and check whether already downloaded organism files are corrupted or not by checking the md5 checksum.
After checking existing files the function will start downloading all remaining organisms.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>reference</th>
      <td><p>a logical value indicating whether or not a genome shall be downloaded if it isn't marked in the database as either a reference genome or a representative genome.</p></td>
    </tr>
    <tr>
      <th>combine</th>
      <td><p>just in case <code>type = "assemblystats"</code> is specified, shall
assemby stats of individual species be imported and combined to a 
<code>data.frame</code>?</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>path to the folder in which downloaded genomes shall be stored. 
By default the kingdom name is used to name the output folder.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a character vector storing the file paths of the retrieved files.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function aims to perform bulk retrieval of the genomes, 
proteomes, cds, etc. of species that belong to the same kingdom of life or 
to the same subgroup.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='meta.retrieval.all.html'>meta.retrieval.all</a></code>, <code><a href='getCollection.html'>getCollection</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># get all available kingdoms for refseq</span>
<span class='fu'><a href='getKingdoms.html'>getKingdoms</a></span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>)
<span class='co'># download all vertebrate genomes from refseq</span>
<span class='fu'>meta.retrieval</span>(<span class='kw'>kingdom</span> <span class='kw'>=</span> <span class='st'>"vertebrate_mammalian"</span>,
               <span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>,
               <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"genome"</span>)

<span class='co'># get all available kingdoms for genbank</span>
<span class='fu'><a href='getKingdoms.html'>getKingdoms</a></span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"genbank"</span>)
<span class='co'># download all vertebrate genomes from genbank</span>
<span class='fu'>meta.retrieval</span>(<span class='kw'>kingdom</span> <span class='kw'>=</span> <span class='st'>"vertebrate_mammalian"</span>,
               <span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"genbank"</span>,
               <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"genome"</span>)

<span class='co'># get all available kingdoms for ensemblgenomes</span>
<span class='fu'><a href='getKingdoms.html'>getKingdoms</a></span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"ensemblgenomes"</span>)
<span class='co'># download all vertebrate genomes from ensemblgenomes</span>
<span class='fu'>meta.retrieval</span>(<span class='kw'>kingdom</span> <span class='kw'>=</span> <span class='st'>"vertebrate_mammalian"</span>,
               <span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"ensemblgenomes"</span>,
               <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"genome"</span>)

<span class='co'># In case users do not wish to retrieve genomes from an entire kingdom, </span>
<span class='co'># but rather from a subgoup (e.g. from species belonging to the </span>
<span class='co'># Gammaproteobacteria class, a subgroup of the bacteria kingdom), </span>
<span class='co'># they can use the following workflow"</span>
<span class='co'># First, users can again consult the getKingdoms() function to retrieve </span>
<span class='co'># kingdom information.</span>
<span class='fu'><a href='getKingdoms.html'>getKingdoms</a></span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>)

<span class='co'># In this example, we will choose the bacteria kingdom. </span>
<span class='co'># Now, the getGroups() function allows users to obtain available </span>
<span class='co'># subgroups of the bacteria kingdom.</span>
<span class='fu'><a href='getGroups.html'>getGroups</a></span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>, <span class='kw'>kingdom</span> <span class='kw'>=</span> <span class='st'>"bacteria"</span>)

<span class='co'># Now we choose the group Gammaproteobacteria and specify </span>
<span class='co'># the group argument in the meta.retrieval() function</span>
<span class='fu'>meta.retrieval</span>(<span class='kw'>kingdom</span> <span class='kw'>=</span> <span class='st'>"bacteria"</span>,
   <span class='kw'>roup</span> <span class='kw'>=</span> <span class='st'>"Gammaproteobacteria"</span>,
   <span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>,
   <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"genome"</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Hajk-Georg Drost

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

